<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-04-27 Mon 23:46 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Inter-individual vs. inter-condition variation in gene expression</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Inter-individual vs. inter-condition variation in gene expression</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org8ea400c">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#orgeebc9f3">Method</a></li>
<li><a href="#org2f68a84">Results</a>
<ul>
<li><a href="#orgcf7b89e">Simplified example</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org8ea400c" class="outline-2">
<h2 id="org8ea400c">Introduction</h2>
<div class="outline-text-2" id="text-org8ea400c">
<p>
Suppose we have gene expression data \(x_{ijk}\) in individuals \(i = 1,
  \ldots, n\), genes \(j = 1, \ldots, p\), and conditions \(k = 1, \ldots,
  m\). The data are noisy observations of true gene expression in each
individual, which in turn are realizations from a population distribution of
gene expression values in each condition. To what extent are inter-condition
differences in gene expression explained by inter-individual differences? \(
  \DeclareMathOperator\N{\mathcal{N}} \DeclareMathOperator\IG{\mathcal{IG}} \)
</p>
</div>
</div>

<div id="outline-container-orgdabeff4" class="outline-2">
<h2 id="setup"><a id="orgdabeff4"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgeebc9f3" class="outline-2">
<h2 id="orgeebc9f3">Method</h2>
<div class="outline-text-2" id="text-orgeebc9f3">
<p>
Suppose \(y_{ijk} = \log_2(x_{ijk} / x_{i+k} + \epsilon)\). Then considering
gene \(j\),
(<a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29">Law
et al. 2014</a>)
</p>

\begin{align}
  y_{ik} \mid \mu_{ik}, s_{ik}^2 &\sim \N(\mu_{ik}, s_{ik}^2)\\
  \mu_{ik} \mid \mu_k, \sigma_k^2 &\sim \N(\mu_k, \sigma_k^2)\\
  \mu_k \mid \mu, \sigma^2 &\sim \N(\mu, \sigma^2).
\end{align}

<p>
In this model, the inter-condition variance is given by \(\sigma^2\) and the
inter-individual variance (within a condition) is given by
\(\sigma_k^2\). We can take an EB-like approach and set \(\mu = \bar{y}\),
which is reasonable because the complete data are informative about the
global mean. Assuming conjugate priors
</p>

\begin{align}
  \sigma_k^2 &\sim \IG(a_k, b_k)\\
  \sigma^2 &\sim \IG(a, b),
\end{align}

<p>
the complete conditionals are analytic
</p>

\begin{align}
  \mu_{ik} \mid \cdot &\sim \N\left(\frac{\sigma_k^2}{s_{ik}^2 + \sigma_k^2} y_{ik} + \frac{s_{ik}^2}{s_{ik}^2 + \sigma_k^2} \mu_{k}, \frac{s_{ik}^2 \sigma_k^2}{s_{ik}^2 + \sigma_k^2} \right)\\
  \sigma_{k}^2 \mid \cdot &\sim \IG\left(a_k + \frac{n}{2}, b_k + \frac{1}{2} \sum_i (\mu_{ik} - \mu_k)^2 \right)\\
  \mu_k \mid \cdot &\sim \N\left(\frac{\sigma^2}{\sigma_k^2 / n + \sigma^2} \frac{1}{n} \sum_i \mu_{ik} + \frac{\sigma_k^2 / n}{\sigma_k^2 / n + \sigma^2} \mu, \frac{\sigma_k^2 \sigma^2 / n}{\sigma_k^2 / n + \sigma^2} \right)\\
  \sigma^2 \mid \cdot &\sim \IG\left(a + \frac{m}{2}, b + \frac{1}{2} \sum_k \mu_{k}^2 \right),
\end{align}

<p>
and we can get posteriors on \(\sigma^2, \sigma_k^2\) by Gibbs sampling.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">est_var</span>(y, s2, ak, bk, a, b, num_samples=1000):
  <span class="org-keyword">assert</span> y.shape == s2.shape
  <span class="org-variable-name">n</span>, <span class="org-variable-name">m</span> = y.shape
  <span class="org-variable-name">mu</span> = y_ik.mean()
  <span class="org-variable-name">sigma_2</span> = 1
  <span class="org-variable-name">mu_k</span> = y_ik.mean(axis=0)
  <span class="org-variable-name">sigma_k_2</span> = np.full(m, 1)
  <span class="org-variable-name">mu_ik</span> = y_ik
  <span class="org-variable-name">samples</span> = [(mu_ik, sigma_k_2, mu_k, sigma_2)]
  <span class="org-variable-name">trace</span> = []
  <span class="org-keyword">for</span> _ <span class="org-keyword">in</span> <span class="org-builtin">range</span>(num_samples):
    <span class="org-variable-name">log_joint</span> = (st.norm(loc=mu_ik, scale=np.sqrt(s2)).logpdf(y).<span class="org-builtin">sum</span>() +
                 st.norm(loc=mu_k, scale=np.sqrt(sigma_k_2)).logpdf(mu_ik).<span class="org-builtin">sum</span>() +
                 st.norm(scale=np.sqrt(sigma_2)).logpdf(mu_k).<span class="org-builtin">sum</span>() +
                 st.invgamma(a=ak + n / 2, scale=bk + .5 * np.square(mu_ik - mu_k).<span class="org-builtin">sum</span>(axis=0)).logpdf(sigma_k_2).<span class="org-builtin">sum</span>() +
                 st.invgamma(a=a + m / 2, scale=b + .5 * np.square(mu_k - mu).<span class="org-builtin">sum</span>()).logpdf(sigma_2))
    trace.append(log_joint)
    <span class="org-comment-delimiter"># </span><span class="org-comment">mu_ik = st.norm(loc=sigma_k_2 / (s2 + sigma_k_2) * y + s2 / (s2 + sigma_k_2) * mu_k, scale=np.sqrt((s2 * sigma_k_2) / (s2 + sigma_k_2))).rvs(size=y.shape)</span>
    <span class="org-variable-name">sigma_k_2</span> = st.invgamma(a=ak + n / 2, scale=bk + .5 * np.square(mu_ik - mu_k).<span class="org-builtin">sum</span>(axis=0)).rvs(size=m)
    <span class="org-variable-name">mu_k</span> = st.norm(loc=sigma_2 / (sigma_k_2 / n + sigma_2) * mu_ik.mean(axis=0) + (sigma_k_2 / n) / (sigma_k_2 / n + sigma_2) * mu, scale=np.sqrt((sigma_k_2 * sigma_2 / n) / (sigma_k_2 / n + sigma_2))).rvs(size=m)
    <span class="org-variable-name">sigma_2</span> = st.invgamma(a=a + m / 2, scale=b + .5 * np.square(mu_k - mu).<span class="org-builtin">sum</span>()).rvs(size=1)
    samples.append((mu_ik, sigma_k_2, mu_k, sigma_2))
  <span class="org-keyword">return</span> samples, trace
</pre>
</div>
</div>
</div>

<div id="outline-container-org2f68a84" class="outline-2">
<h2 id="org2f68a84">Results</h2>
<div class="outline-text-2" id="text-org2f68a84">
</div>
<div id="outline-container-orgcf7b89e" class="outline-3">
<h3 id="orgcf7b89e">Simplified example</h3>
<div class="outline-text-3" id="text-orgcf7b89e">
<p>
Simulate some data.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(1)
<span class="org-variable-name">m</span> = 2
<span class="org-variable-name">n</span> = 100
<span class="org-variable-name">s2</span> = 1
<span class="org-variable-name">mu_k</span> = np.random.normal(scale=np.sqrt(s2), size=m)
<span class="org-variable-name">s2_k</span> = np.full(m, .5)
<span class="org-variable-name">mu_ik</span> = np.random.normal(loc=mu_k, scale=np.sqrt(s2_k), size=(n, m))
<span class="org-variable-name">s2_ik</span> = np.full((n, m), .01)
<span class="org-variable-name">y_ik</span> = np.random.normal(loc=mu_ik, scale=np.sqrt(s2_ik))
</pre>
</div>

<p>
Plot the simulated example.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
plt.gcf().set_size_inches(2.5, 2.5)
<span class="org-variable-name">grid</span> = np.linspace(-2, 5, 1000)
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> <span class="org-builtin">range</span>(m):
  plt.hist(y_ik[:,k], bins=12, density=<span class="org-constant">True</span>, color=cm(k), alpha=0.5)
  plt.plot(grid, st.norm(loc=mu_k[k], scale=np.sqrt(s2_k[k])).pdf(grid), color=cm(k))
plt.xlabel(<span class="org-string">'$y$'</span>)
plt.ylabel(<span class="org-string">'Density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/de.org/sim-ex.png" alt="sim-ex.png">
</p>
</div>

<p>
Get posterior samples.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(2)
np.seterr(<span class="org-string">'raise'</span>)
<span class="org-variable-name">samples</span>, <span class="org-variable-name">trace</span> = est_var(y_ik, s2_ik, ak=1e-2, bk=1e-2, a=1e-2, b=1e-2, num_samples=2000)
<span class="org-variable-name">samples</span> = samples[-1000:]
</pre>
</div>

<p>
Plot the log joint over the samples.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3.5, 2)
plt.plot(trace, lw=1, c=<span class="org-string">'k'</span>)
plt.xlabel(<span class="org-string">'Sample'</span>)
plt.ylabel(<span class="org-string">'Log joint'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/de.org/sim-trace.png" alt="sim-trace.png">
</p>
</div>

<p>
Plot the estimated posterior distributions.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 3)
fig.set_size_inches(7.5, 2.5)
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> <span class="org-builtin">range</span>(m):
  <span class="org-variable-name">query</span> = np.log(np.array([s[1][k] <span class="org-keyword">for</span> s <span class="org-keyword">in</span> samples]))
  ax[k].hist(query, bins=10, density=<span class="org-constant">True</span>, color=<span class="org-string">'0.7'</span>)
  ax[k].axvline(x=np.log(s2_k[k]), c=<span class="org-string">'k'</span>, lw=1)
  ax[k].axvline(x=query.mean(), c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">'--'</span>)
  ax[k].set_xlabel(rf<span class="org-string">'$\ln(\sigma_{k}^2)$'</span>)

<span class="org-variable-name">query</span> = np.log(np.array([s[-1] <span class="org-keyword">for</span> s <span class="org-keyword">in</span> samples]))
ax[2].hist(query, bins=10, density=<span class="org-constant">True</span>, color=<span class="org-string">'0.7'</span>)
ax[2].set_xlabel(r<span class="org-string">'$\ln(\sigma^2)$'</span>)
ax[2].axvline(x=np.log(s2), c=<span class="org-string">'k'</span>, lw=1)
ax[2].axvline(x=query.mean(), c=<span class="org-string">'r'</span>, lw=1, ls=<span class="org-string">'--'</span>)
ax[0].set_ylabel(<span class="org-string">'Density'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/de.org/sim-post.png" alt="sim-post.png">
</p>
</div>

<p>
Get the posterior distribution of \(\sigma_k^2 / \sigma^2\).
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(2.5, 2.5)
<span class="org-variable-name">query</span> = np.array([s[1][0] <span class="org-keyword">for</span> s <span class="org-keyword">in</span> samples]) / np.array([s[-1] <span class="org-keyword">for</span> s <span class="org-keyword">in</span> samples]).ravel()
plt.hist(query, bins=10, density=<span class="org-constant">True</span>, color=<span class="org-string">'0.7'</span>)
plt.axvline(x=s2_k[0] / s2, lw=1, c=<span class="org-string">'k'</span>)
plt.axvline(x=query.mean(), lw=1, ls=<span class="org-string">'--'</span>, c=<span class="org-string">'r'</span>)
plt.xlabel(r<span class="org-string">'$\sigma_0^2 / \sigma^2$'</span>)
plt.ylabel(<span class="org-string">'Density'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/de.org/sim-post-ratio.png" alt="sim-post-ratio.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-04-27 Mon 23:46</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
