<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2021-01-08 Fri 12:06 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Experimental fine-mapping of QTLs from scRNA-seq data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Experimental fine-mapping of QTLs from scRNA-seq data</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5f78797">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org8ef923c">Method</a>
<ul>
<li><a href="#org67a759b">Regulatory network inference from scRNA-seq data</a></li>
</ul>
</li>
<li><a href="#orge254607">Results</a>
<ul>
<li><a href="#org56ea0c7">Proof of concept</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org5f78797" class="outline-2">
<h2 id="org5f78797">Introduction</h2>
<div class="outline-text-2" id="text-org5f78797">
<p>
One of the fundamental problems in translating genetic associations with
disease into actionable insights is identifying the molecular mechanism by
which they operate. What do we mean by &ldquo;identifying a molecular mechanism&rdquo;?
Ideally, we would identify a cell type of action, a collection of regulatory
elements, a target gene, and collection of upstream regulators (e.g.,
Smemo et al. 2013, Claussnitzer et al. 2015, Sekar et al. 2016). 
</p>

<p>
Considerable progress has been made on identifying genomic annotations
enriched for associations, which can potentially be used to identify putative
cell types of action (e.g., Maurano et al. 2012, Gusev et al. 2015, Finucane
et al. 2015, Finucane et al. 2016). Progress has also been made on
statistically identifying putative causal variants directly from genotype and
phenotype, possibly using enrichments as prior information (e.g., Hormozdiari
et al. 2014, Pickrell 2014, Kichaev et al. 2014, Hormozdiari et al. 2014,
Chung et al. 2014, Li and Kellis 2016). Recent attention has focused on the
statistical problem of identifying the target gene from genotype, phenotype,
and gene expression data (e.g., Gamazon et al. 2015, Gusev et al. 2016,
Mancuso et al. 2017, Zhu et al. 2018).  
</p>

<p>
Together, these statistical approaches can potentially be combined to
identify a putative mechanism by which a disease-associated locus causes
disease risk/onset. However, they are fundamentally limited by the
availability of cell type&#x2013;specific molecular data at the transcriptomic and
epigenomic levels, which are required to derive relevant genomic
annotations. These limitations have, in part, motivated recent efforts to
generate relevant single cell genomic data (e.g., Regev et al. 2017).
</p>

<p>
In parallel, experimental approaches have been developed to identify
functional regulatory elements using transfected reporter constructs (e.g.,
Melkinov et al. 2012, Arnold et al. 2013, Tewhey et al. 2016, Ernst et
al. 2016, Wang et al. 2018), or more recently, combining CRISPR-perturbation
with scRNA-seq phenotyping (e.g., Adamson et al. 2016, Dixit et al. 2016,
Jaitin et al. 2016, Xie et al. 2017, Datlinger et al. 2017). These approaches
have been limited by the cell type in which experiments are performed.
</p>

<p>
Here, we outline an alternative approach, which exploits the fact that
naturally occuring stochastic variation in protein abundances of upstream
regulators will lead to downstream variation in mRNA levels of target
genes. This fact suggests that it could possible to identify upstream
regulators, functional regulatory elements, and target genes directly from
scRNA-seq data. Such an approach could avoid problems of collecting large
numbers of donor individuals, and of picking the correct cell
type/developmental time point <i>a priori</i>. Further, it could more directly
answer the relevant questions for interpreting disease-associated
loci. Longer term, this methodology could be extended to spatial
transcriptomic measurements, allowing us to account for microenvironments and
cell-cell interactions.
</p>
</div>
</div>

<div id="outline-container-org8b51cf7" class="outline-2">
<h2 id="setup"><a id="org8b51cf7"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
<span class="org-keyword">import</span> txpred.models.susie
<span class="org-keyword">import</span> torch
<span class="org-keyword">import</span> torch.utils.data <span class="org-keyword">as</span> td

rpy2.robjects.pandas2ri.activate()
<span class="org-variable-name">mrash</span> = rpy2.robjects.packages.importr(<span class="org-string">'mr.ash.alpha'</span>)
<span class="org-variable-name">susie</span> = rpy2.robjects.packages.importr(<span class="org-string">'susieR'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org8ef923c" class="outline-2">
<h2 id="org8ef923c">Method</h2>
<div class="outline-text-2" id="text-org8ef923c">
</div>
<div id="outline-container-org67a759b" class="outline-3">
<h3 id="org67a759b">Regulatory network inference from scRNA-seq data</h3>
<div class="outline-text-3" id="text-org67a759b">
<p>
We assume (Sarkar and Stephens 2020) \(
   \DeclareMathOperator\Pois{Poisson}
   \DeclareMathOperator\N{\mathcal{N}}
   \newcommand\xiplus{x_{i+}}
   \newcommand\mh{\mathbf{H}}
   \newcommand\vb{\mathbf{b}}
   \newcommand\mi{\mathbf{I}}
   \)
</p>

\begin{equation}
  x_{ij} \mid \xiplus, \eta_{ij} \sim \Pois(\xiplus \exp(\eta_{ij})),
\end{equation}

<p>
where \(x_{ij}\) denotes the number of molecules of gene \(j\) observed in
cell \(i\) and \(\xiplus \triangleq \sum_j x_{ij}\). We further assume
</p>

\begin{align}
  \eta_{\cdot j} &\sim \N([\mh]_{\cdot j} \vb_j, \sigma^2 \mi)\\
  b_{j j^{\prime}} &\sim \pi \N(0, \sigma^2 \sigma_b^2) + (1 - \pi) \delta_0(\cdot),
\end{align}

<p>
where \(\mh = [\eta_{ij}]\) and \([\mh]_{ij}\) denotes the submatrix of
\(\mh\) obtained by deleting the \(i\)th row and \(j\)th column. Then, the
vectors \(\vb_j\) give the off-diagonal elements of the adjacency matrix of
the gene-gene-interaction network. We can estimate \(\eta_{ij}\) from the
data using e.g., <a href="mpebpm.html">mpebpm</a> (Sarkar et al. 2019), and
estimate \(\vb_j\) from the data using e.g., <code>susie</code> (Wang et al. 2018).
</p>
</div>
</div>
</div>

<div id="outline-container-orge254607" class="outline-2">
<h2 id="orge254607">Results</h2>
<div class="outline-text-2" id="text-orge254607">
</div>
<div id="outline-container-org56ea0c7" class="outline-3">
<h3 id="org56ea0c7">Proof of concept</h3>
<div class="outline-text-3" id="text-org56ea0c7">
<p>
Read the iPSC data (Sarkar et al. 2019).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/ipsc/ipsc.h5ad'</span>)
</pre>
</div>

<p>
Read the estimated expression models.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">log_mu</span> = np.load(<span class="org-string">'/scratch/midway2/aksarkar/ideas/mpebpm-ipsc-design-log-mu.npy'</span>)
<span class="org-variable-name">neg_log_phi</span> = np.load(<span class="org-string">'/scratch/midway2/aksarkar/ideas/mpebpm-ipsc-design-neg-log-phi.npy'</span>)
<span class="org-variable-name">logodds</span> = np.load(<span class="org-string">'/scratch/midway2/aksarkar/ideas/mpebpm-ipsc-design-logodds.npy'</span>)
</pre>
</div>

<p>
Compute \(E[\ln\lambda_{ij}]\) to zeroth order. (This is needed because the
expectation is not finite when \(\pi_j > 0\).)
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">onehot</span> = pd.get_dummies(dat.obs[<span class="org-string">'chip_id'</span>], sparse=<span class="org-constant">True</span>)
<span class="org-keyword">del</span> onehot[<span class="org-string">'NA18498'</span>]
<span class="org-variable-name">onehot</span> = onehot.sparse.to_coo().tocsr()
<span class="org-comment-delimiter"># </span><span class="org-comment">Important: this is cells by genes, not donors by genes</span>
<span class="org-variable-name">log_mean</span> = (onehot @ -np.log1p(-logodds)) + np.log1p(dat.X + onehot @ np.exp(neg_log_phi)) - np.log(dat.X.<span class="org-builtin">sum</span>(axis=1) + onehot @ np.exp(-log_mu + neg_log_phi))
</pre>
</div>

<p>
Download a <a href="http://humantfs.ccbr.utoronto.ca/download.php">curated
database</a> of TF genes
(<a href="https://doi.org/10.1016/j.cell.2018.01.029">Lambert et al. 2018</a>)
</p>

<div class="org-src-container">
<pre class="src src-sh">curl -OL <span class="org-string">"http://humantfs.ccbr.utoronto.ca/download/v_1.01/DatabaseExtract_v_1.01.csv"</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">tfs</span> = pd.read_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/DatabaseExtract_v_1.01.csv'</span>, index_col=0)
<span class="org-variable-name">query</span> = dat.var.merge(tfs.loc[tfs[<span class="org-string">'Is TF?'</span>] == <span class="org-string">'Yes'</span>], left_index=<span class="org-constant">True</span>, right_on=<span class="org-string">'Ensembl ID'</span>, how=<span class="org-string">'inner'</span>).index
<span class="org-variable-name">design</span> = log_mean[:,query]
</pre>
</div>

<p>
Focus on <i>SKP1</i>, an eQTL found in the previous study. First, fit <code>sgvbvs</code>,
and report genes with non-trivial PIP.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">gene</span> = <span class="org-string">'SKP1'</span>
<span class="org-variable-name">batch_size</span> = 128
<span class="org-variable-name">seed</span> = 3
torch.manual_seed(seed)
<span class="org-variable-name">j</span> = np.where(dat.var[<span class="org-string">'name'</span>] == gene)[0][0]
<span class="org-variable-name">data</span> = td.DataLoader(
  td.TensorDataset(
    torch.tensor(design),
    torch.tensor(log_mean[:,j])),
  batch_size=batch_size,
  shuffle=<span class="org-constant">True</span>)
<span class="org-variable-name">fit0</span> = (txpred.models.susie.GaussianRegression(txpred.models.susie.SpikeSlab(design.shape[1]))
        .fit(data, n_epochs=100, log_dir=f<span class="org-string">'runs/txpred/{gene}-{batch_size}-{seed}'</span>))
dat[:,query].var.loc[(torch.sigmoid(fit0.prior.logits) &gt; 0.1).numpy()]
</pre>
</div>

<pre class="example">
chr     start       end name strand      source
index
ENSG00000087086  hs19  49468558  49470135  FTL      +  H. sapiens
</pre>

<p>
Fit <code>susie</code> to get credible sets.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">fit</span> = susie.susie(design, log_mean[:,j], L=5)
fit.rx2(<span class="org-string">'V'</span>)
</pre>
</div>

<pre class="example">
array([0.04996618, 0.0089835 , 0.00727507, 0.00513533, 0.00376015])

</pre>

<p>
Get the credible sets.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cs</span> = [susie.susie_get_cs(fit, coverage=0.95).rx2(<span class="org-string">'cs'</span>).rx2(i + 1) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(5)]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">dat[:,query].var.iloc[np.array(cs).ravel()]
</pre>
</div>

<pre class="example">
chr     start       end     name strand      source
index
ENSG00000060656   hs1  29563028  29653325    PTPRU      +  H. sapiens
ENSG00000103145  hs16   3072621   3074287  HCFC1R1      -  H. sapiens
ENSG00000054116   hs1  36602173  36615098  TRAPPC3      -  H. sapiens
ENSG00000059122  hs16   2961938   3001209  FLYWCH1      +  H. sapiens
ENSG00000083544  hs13  60970591  61148012    TDRD3      +  H. sapiens
</pre>

<p>
Get all predicted TF binding motifs within 1MB of <i>SKP1</i>.
</p>

<div class="org-src-container">
<pre class="src src-sh">module load htslib
curl -OL <span class="org-string">"http://compbio.mit.edu/encode-motifs/enrichments.txt.gz"</span>
curl -OL <span class="org-string">"http://compbio.mit.edu/encode-motifs/matches.txt.gz"</span>
zcat matches.txt.gz | sort -k2,2 -k3,3n -k4,4n | bgzip &gt;motif-matches.txt.gz
tabix -s 2 -b 3 -e 4 motif-matches.txt.gz
</pre>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2021-01-08 Fri 12:06</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
