<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2021-04-30 Fri 13:52 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Learning the structure of RNA expression variation</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Learning the structure of RNA expression variation</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5fc0aab">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#orgfa22f8e">Main results</a>
<ul>
<li><a href="#orge4e6819">Pre-defined conditions/hard clustering</a></li>
<li><a href="#org3143bf8">Generalizing to soft clustering</a></li>
<li><a href="#org66c8c51">Generalizing to a non-linear latent variable model</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org5fc0aab" class="outline-2">
<h2 id="org5fc0aab">Introduction</h2>
<div class="outline-text-2" id="text-org5fc0aab">
<p>
Characterizing the extent of RNA expression variation within cell types, cell
states, donors, and experimental conditions is important to learn about
evolutionary constraint on gene regulation, modulation of differentiation
potential, and the mechanisms underlying cell fate decisions. However,
actually estimating the level of expression variation from scRNA-seq data is
difficult. Here, we review some of these difficulties. \(
  \DeclareMathOperator\Gam{Gamma}
  \DeclareMathOperator\N{\mathcal{N}}
  \DeclareMathOperator\Pois{Poisson}
  \DeclareMathOperator\V{V}
  \newcommand\mf{\mathbf{F}}
  \newcommand\mh{\mathbf{H}}
  \newcommand\mi{\mathbf{I}}
  \newcommand\ml{\mathbf{L}}
  \newcommand\vl{\mathbf{l}}
  \newcommand\vx{\mathbf{x}}
  \newcommand\vz{\mathbf{z}}
  \newcommand\vmu{\boldsymbol{\mu}}
  \newcommand\vphi{\boldsymbol{\phi}}
  \newcommand\vpi{\boldsymbol{\pi}}
  \)
</p>
</div>
</div>

<div id="outline-container-org31cd25d" class="outline-2">
<h2 id="setup"><a id="org31cd25d"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> mpebpm.topics
<span class="org-keyword">import</span> scanpy <span class="org-keyword">as</span> sc
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> torch
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgfa22f8e" class="outline-2">
<h2 id="orgfa22f8e">Main results</h2>
<div class="outline-text-2" id="text-orgfa22f8e">
</div>
<div id="outline-container-orge4e6819" class="outline-3">
<h3 id="orge4e6819">Pre-defined conditions/hard clustering</h3>
<div class="outline-text-3" id="text-orge4e6819">
<p>
The simplest case to consider is analyzing scRNA-seq observations
from a single gene in one condition. Then, one can fit an observation model
(Sarkar and Stephens 2020)
</p>

\begin{align}
  x_i \mid s_i, \lambda_i &\sim \Pois(s_i \lambda_i), \qquad i = 1, \ldots, n\\
  \lambda_i &\sim g = \Gam(\phi^{-1}, \mu^{-1}\phi^{-1}).
\end{align}

<p>
Under this model, it is clear that the variance of true gene expression
levels is \(\V[g] = \mu^2\phi\). More generally, one could instead assume
that \(g\) belongs to some other family, but still characterize expression
variance as \(\V[g]\). There is some evidence that families more flexible
than Gamma distributions may be required for a large fraction of genes in
some data sets (Sarkar and Stephens 2020). One goal of this analysis is to
potentially interpret gene expression variance relative to the mean as
indicative of
<a href="diff-var.html#org5d82bf1">the
underlying kinetic parameters of transcriptional bursting</a>.
</p>

<p>
In the case where \(K\) conditions were pre-defined (e.g., donor) or are
learned from the data by hard-clustering, one can readily generalize the
above observation model to multiple genes and conditions (Sarkar et
al. 2019)
</p>

\begin{align}
  x_{ij} \mid s_i, \lambda_{ij} &\sim \Pois(s_i \lambda_{ij}), \qquad i = 1, \ldots, n; j = 1, \ldots, p\\
  \lambda_{ij} &= \mu_{ij} u_{ij}\\
  \mu_{ij} &= \exp((\ml\mf')_{ij})\\
  u_{ij} \mid \ml, \mh &\sim \Gam(\exp(-(\ml\mh')_{ij}), \exp(-(\ml\mh')_{ij})),
\end{align}

<p>
where \(\ml\) is an \(n \times K\) fixed, binary matrix corresponding to
assignment of samples to conditions, \(\mf\) is a \(K \times p\) matrix of
mean parameters, and \(\mh\) is a \(K \times p\) matrix of dispersion
parameters. This observation model can be readily fit <a href="mpebpm.html">to
massive data using stochastic gradient descent</a>. Marginalizing,
</p>

\begin{equation}
  \lambda_{ij} \sim g_{ij}(\cdot) = \Gam(\exp(-(\ml\mh')_{ij}), \exp(-(\ml(\mf + \mh)')_{ij})),
\end{equation}

<p>
where \(g_{ij}\) is identical for observations from the same condition and
gene. Therefore, the variance of gene expression can still be estimated as
the distinct \(\V[g_{ij}]\). However, this approach has important
limitations: (1) it requires pre-defining conditions, which may not be
justified by the data (in the case of learning clusters), and (2) it treats
every gene independently, and does not consider propagation of noise in gene
regulatory networks (e.g.,
<a href="https://www.pnas.org/content/98/15/8614">Thattai and van Oudenaarden
2001</a>).
</p>
</div>
</div>

<div id="outline-container-org3143bf8" class="outline-3">
<h3 id="org3143bf8">Generalizing to soft clustering</h3>
<div class="outline-text-3" id="text-org3143bf8">
<p>
Considering the expression model
</p>

\begin{equation}
  \lambda_{ij} \mid \ml, \mf, \mh \sim \Gam(\exp(-(\ml\mh')_{ij}), \exp(-(\ml(\mf + \mh)')_{ij})),
\end{equation}

<p>
what happens when \(\ml\) is not binary? Suppose \(\ml\) is normalized so
that its rows sum to 1; then, this model corresponds to a sort of &ldquo;noisy
topic model,&rdquo; where topics are characterized by a collection of Gamma
distributions for each gene, and cells are convex combinations of those
topics. In some sense then, the topics in this model could be thought of as
&ldquo;noisy gene expression programs&rdquo;. One can in principle fit this model by
automatic differentiation and gradient descent, ensuring the sum constraint
by re-parameterizing \(\vl_i = \operatorname{softmax}(\tilde{\vl}_i)\). One
could also amortize inference by introducing an inference network mapping
\(\vx_i\) to \(\vl_i\).
</p>

<p>
<i>Remark.</i> When \(\ml\) is binary, using the exponential inverse link
function makes sense since we are selecting a single log mean parameter and
a single log dispersion parameter for each observation. However, when
\(\ml\) is instead normalized to have row sum 1, it is not clear that this
model makes sense compared to a model with identity link, analogous to
NMF/pLSI.
</p>

<p>
This approach addresses the issues that pre-defined conditions are not
always available, and hard-clustering data may not be justified. However, it
complicates the question of what precisely the variance of gene expression
corresponds to in the model. If we think of samples as being mixtures of
gene expression programs, then even the question of what &ldquo;variance of gene
expression&rdquo; means is problematic, since it is no longer clear what subset of
samples the variance is being computed over. Is the variance of each learned
gene expression program instead the relevant quantity, regardless of which
samples express it? Is this the correct way to think about asking the
question when we don&rsquo;t begin from the assumption that there were some number
of distinct cell types present?
</p>

<p>
In this model, we can think of \(\vl_i\) as a latent variable, and the model
above as a linear latent variable model. But now, expression variation has
components \(\V[\vl_i]\) and \(\V[\lambda_{ij} \mid \vl_i]\). Which of these
components corresponds to expression noise? What does it mean for
\(\V[\lambda_{ij} \mid \vl_i] \neq 0\) when \(\vl_i\) is a latent variable
for a single observation? When \(\vl_i\) is constrained to lie in the
simplex, what does \(\V[\vl_i]\) even mean?
</p>

<p>
<i>Remark.</i> This model is a generalization of <a href="nbmix.html">a simple
model-based clustering method</a>, in which hard clusters are defined by a
collection of Gamma distributions
</p>

\begin{equation}
  \lambda_{ij} \mid \vpi_i, \vmu_k, \vphi_k \sim \sum_{k=1}^{K} \pi_{ik} \Gam(\phi_{kj}^{-1}, \phi_{kj}^{-1}\mu_{kj}^{-1}).
\end{equation}

<p>
It also generalizes other methods
(<a href="https://bmcsystbiol.biomedcentral.com/articles/10.1186/s12918-019-0699-6">scNBMF</a>,
<a href="https://academic.oup.com/bioinformatics/article/36/11/3418/5807606">LDVAE</a>),
which fit expression models of the form
</p>

\begin{align}
  \lambda_{ij} &= \mu_{ij} u_{ij}\\
  \mu_{ij} &= \exp((\ml\mf')_{ij})\\
  u_{ij} \mid \phi_j &\sim \Gam(\phi_j^{-1}, \phi_j^{-1}),
\end{align}

<p>
i.e., where \(p(u_{ij})\) is constant across conditions. It is also a
related to <a href="https://doi.org/10.1038/s41467-018-07931-2">DCA</a>, which fits a
similar expression model, but does not constrain \(\vl_i\) to be in the
simplex and assumes a non-linear mapping from \(\vl_i\) to mean and
dispersion parameters per cell per gene .
</p>
</div>
</div>

<div id="outline-container-org66c8c51" class="outline-3">
<h3 id="org66c8c51">Generalizing to a non-linear latent variable model</h3>
<div class="outline-text-3" id="text-org66c8c51">
<p>
Now consider the non-linear latent variable model
</p>

\begin{align}
  x_{ij} \mid s_i, \lambda_{ij} &\sim \Pois(s_i \lambda_{ij})\\
  \lambda_{ij} &= (f(\vz_i))_j\\
  \vz_i &\sim \N(0, \mi).
\end{align}

<p>
Here, \(f\) is a fully-connected neural network, and variability is pushed
into the \(\vz_i\). In this model, expression noise is completely explained
by \(\V[\vz_i \mid \vx_i]\). That is, variation in expression levels is
completely explained by variation in a low-dimensional latent space. Is this
a reasonable model? One thing that isn&rsquo;t obvious is what exactly expression
variance corresponds to in this model, since it is not clear which subset of
samples the variance is being computed over. Naively, one could return to
hard-clustering the \(\vz_i\), but it is not clear that this is an
improvement over existing hard-clustering approaches.
</p>

<p>
Existing methods (<a href="https://doi.org/10.1038/s41592-018-0229-2">scVI</a>,
<a href="https://www.nature.com/articles/s41592-020-01050-x">totalVI</a>) fit an even
more complex expression model
</p>

\begin{align}
  \lambda_{ij} &= \mu_{ij} u_{ij}\\
  \mu_{ij} &= (f(\vz_i))_j\\
  u_{ij} &\sim \Gam(\phi_j^{-1}, \phi_j^{-1})\\
  \vz_i &\sim \N(0, \mi).
\end{align}

<p>
In this model, expression noise depends both on variation in the latent
space and extra variability. Does this make sense?
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2021-04-30 Fri 13:52</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
