<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-06-22 Mon 14:20 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Grand challenges in single-cell data science</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Grand challenges in single-cell data science</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgfe92fa2">Introduction</a></li>
<li><a href="#orgc310edd">Recurring challenges</a>
<ul>
<li><a href="#org29bcd62">Varying levels of resolution</a></li>
<li><a href="#org412686e">Quantifying uncertainty of measurements and analysis results</a></li>
<li><a href="#org0bfd6b8">Scaling to higher dimensionalities: more cells, more features, and broader coverage</a></li>
</ul>
</li>
<li><a href="#orgcf2a286">Challenges in single-cell transcriptomics</a>
<ul>
<li><a href="#org25f7f2e">Handling sparsity in single-cell RNA sequencing</a></li>
<li><a href="#orgdf8c25b">Defining flexible statistical frameworks for discovering complex differential patterns in gene expression</a></li>
<li><a href="#org753a584">Mapping single cells to a reference atlas</a></li>
<li><a href="#org9e0a234">Generalizing trajectory inference</a></li>
<li><a href="#org85d4405">Finding patterns in spatially resolved measurements</a></li>
</ul>
</li>
<li><a href="#org2bf77d5">Challenges in single-cell genomics</a>
<ul>
<li><a href="#org1701f2f">Dealing with errors and missing data in the identification of variation from single-cell DNA sequencing data</a></li>
</ul>
</li>
<li><a href="#orge9a2016">Challenges in single-cell phylogenomics</a>
<ul>
<li><a href="#orga37ea44">Scaling phylogenetic models to many cells and many sites</a></li>
<li><a href="#org1d6d0de">Integrating multiple types of variation into phylogenetic models</a></li>
<li><a href="#org793fc14">Inferring population genetic parameters of tumor heterogeneity by model integration</a></li>
</ul>
</li>
<li><a href="#org7c4f07b">Overarching challenges</a>
<ul>
<li><a href="#orged12648">Integration of single-cell data across samples, experiments, and types of measurement</a></li>
<li><a href="#org50e6c39">Validating and benchmarking analysis tools for single-cell measurements</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orgfe92fa2" class="outline-2">
<h2 id="orgfe92fa2">Introduction</h2>
<div class="outline-text-2" id="text-orgfe92fa2">
<p>
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1926-6">Lähnemann
et al. 2020</a> list 11 &ldquo;grand challenges&rdquo; in single cell data science. On
first reading, the list of challenges struck me as reviewing areas where
researchers have already made the fundamental contributions, rather than
areas where fundamental contributions remain to be made. Here, I revisit the
paper more carefully to try and identify areas where fundamental
contributions could be made.
</p>
</div>
</div>

<div id="outline-container-orgc310edd" class="outline-2">
<h2 id="orgc310edd">Recurring challenges</h2>
<div class="outline-text-2" id="text-orgc310edd">
</div>
<div id="outline-container-org29bcd62" class="outline-3">
<h3 id="org29bcd62">Varying levels of resolution</h3>
</div>

<div id="outline-container-org412686e" class="outline-3">
<h3 id="org412686e">Quantifying uncertainty of measurements and analysis results</h3>
</div>

<div id="outline-container-org0bfd6b8" class="outline-3">
<h3 id="org0bfd6b8">Scaling to higher dimensionalities: more cells, more features, and broader coverage</h3>
</div>
</div>

<div id="outline-container-orgcf2a286" class="outline-2">
<h2 id="orgcf2a286">Challenges in single-cell transcriptomics</h2>
<div class="outline-text-2" id="text-orgcf2a286">
</div>
<div id="outline-container-org25f7f2e" class="outline-3">
<h3 id="org25f7f2e">Handling sparsity in single-cell RNA sequencing</h3>
</div>

<div id="outline-container-orgdf8c25b" class="outline-3">
<h3 id="orgdf8c25b">Defining flexible statistical frameworks for discovering complex differential patterns in gene expression</h3>
</div>

<div id="outline-container-org753a584" class="outline-3">
<h3 id="org753a584">Mapping single cells to a reference atlas</h3>
</div>

<div id="outline-container-org9e0a234" class="outline-3">
<h3 id="org9e0a234">Generalizing trajectory inference</h3>
</div>

<div id="outline-container-org85d4405" class="outline-3">
<h3 id="org85d4405">Finding patterns in spatially resolved measurements</h3>
</div>
</div>

<div id="outline-container-org2bf77d5" class="outline-2">
<h2 id="org2bf77d5">Challenges in single-cell genomics</h2>
<div class="outline-text-2" id="text-org2bf77d5">
</div>
<div id="outline-container-org1701f2f" class="outline-3">
<h3 id="org1701f2f">Dealing with errors and missing data in the identification of variation from single-cell DNA sequencing data</h3>
</div>
</div>

<div id="outline-container-orge9a2016" class="outline-2">
<h2 id="orge9a2016">Challenges in single-cell phylogenomics</h2>
<div class="outline-text-2" id="text-orge9a2016">
</div>
<div id="outline-container-orga37ea44" class="outline-3">
<h3 id="orga37ea44">Scaling phylogenetic models to many cells and many sites</h3>
</div>

<div id="outline-container-org1d6d0de" class="outline-3">
<h3 id="org1d6d0de">Integrating multiple types of variation into phylogenetic models</h3>
</div>

<div id="outline-container-org793fc14" class="outline-3">
<h3 id="org793fc14">Inferring population genetic parameters of tumor heterogeneity by model integration</h3>
</div>
</div>

<div id="outline-container-org7c4f07b" class="outline-2">
<h2 id="org7c4f07b">Overarching challenges</h2>
<div class="outline-text-2" id="text-org7c4f07b">
</div>
<div id="outline-container-orged12648" class="outline-3">
<h3 id="orged12648">Integration of single-cell data across samples, experiments, and types of measurement</h3>
</div>

<div id="outline-container-org50e6c39" class="outline-3">
<h3 id="org50e6c39">Validating and benchmarking analysis tools for single-cell measurements</h3>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-06-22 Mon 14:20</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
