<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2021-03-25 Thu 18:48 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Grand challenges in single-cell data science</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Grand challenges in single-cell data science</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgfe92fa2">Introduction</a></li>
<li><a href="#orgc310edd">Recurring challenges</a></li>
<li><a href="#orgcf2a286">Challenges in single-cell transcriptomics</a></li>
<li><a href="#org2bf77d5">Challenges in single-cell genomics</a></li>
<li><a href="#orge9a2016">Challenges in single-cell phylogenomics</a></li>
<li><a href="#org7c4f07b">Overarching challenges</a></li>
</ul>
</div>
</div>

<div id="outline-container-orgfe92fa2" class="outline-2">
<h2 id="orgfe92fa2">Introduction</h2>
<div class="outline-text-2" id="text-orgfe92fa2">
<p>
<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-1926-6">LÃ¤hnemann
et al. 2020</a> list 11 &ldquo;grand challenges&rdquo; in single cell data science. On
first reading, the list of challenges struck me as reviewing areas where
researchers have already made the fundamental contributions, rather than
areas where fundamental contributions remain to be made. Here, I revisit the
paper more carefully to try and identify areas where fundamental
contributions could be made.
</p>
</div>
</div>

<div id="outline-container-orgc310edd" class="outline-2">
<h2 id="orgc310edd">Recurring challenges</h2>
<div class="outline-text-2" id="text-orgc310edd">
<p>
<b>Varying levels of resolution.</b> The discussion in this section is initially
motivated by the problem of unsupervised learning of clusters in the context
of scRNA-seq data, although the overarching questions are more general. The
overarching questions (which remain unstated explicitly, but implicitly
motivate which considerations are deemed &ldquo;important&rdquo;) are: (1) how can we
identify subsets of cells that make sense to treat as coherent units, (2) how
can we choose the &ldquo;resolution&rdquo; (degree of coherence on the axis of interest
versus degree of heterogeneity elsewhere) for a particular scientific
question of interest, and (3) how do we build representations of the data
that facilitate extracting coherent subsets of cells at the desired
resolution?
</p>

<p>
The highlighted example of PAGA
(<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1663-x">Wolf
et al. 2019</a>) is illustrative of the assumptions that went into the framing
of the problem. The example suggests that the challenge should be addressed
by building monolithic, hierarchical representations that capture multiple
levels of resolution at once, such that exploratory analysis becomes
trivial. However, other methods such as HSNE
(<a href="https://onlinelibrary.wiley.com/doi/full/10.1111/cgf.12878">Pezotti et
al. 2016</a>) or topic models
(<a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006599">Dey
et al. 2017</a>,
<a href="https://www.nature.com/articles/s41592-019-0367-1">GonzÃ¡lez-Blas et
al. 2019</a>) require a different approach, where higher resolution
representations require both a human in the loop, and potentially substantial
extra computation.
</p>

<p>
Building monolithic, hierarchical representations requires developing new
statistical models with (likely) complicated inference algorithms. It is not
immediately obvious whether such representations would be useful outside of
biology, and if so in what way. In contrast, building high quality, efficient
software implementations of relatively simple methods could be a much more
feasible strategy to make exploratory data analysis at multiple levels of
resolution easier. There are two potential objections to the latter strategy:
(1) it allows researcher degrees of freedom
(<a href="https://osf.io/n3axs/download">Gelman and Loken 2013</a>), and (2) it is
difficult to assess uncertainty in the estimated hierarchy. These can be
countered by the fact that the methods are used for exploratory data
analysis, not confirmatory data analysis
(<a href="http://theta.edu.pl/wp-content/uploads/2012/10/exploratorydataanalysis_tukey.pdf">Tukey
1977</a>), and that in many cases the estimated representations and hierarchies
can be confirmed by external, independent data.
</p>

<p>
<b>Quantifying uncertainty of measurements and analysis results.</b> There are two
issues: (1) the measurement noise in single cell assays is increased compared
to bulk assays, making certain analyses harder/impossible; (2) common
analyses have uncertainties that are often not propagated through typical
analysis pipelines. The first issue is not properly appreciated in the field,
but unfortunately, likely cannot be addressed through clever computation. For
example, accurately estimating gene expression variance from scRNA-seq data
appears generally not possible. 
</p>

<p>
The second issue initially appears more difficult. What does a typical
analysis pipeline actually do? (1) filter e.g., low quality cells, and
uninformative genes, (2) estimate a low rank representation, which induces a
visualization and a clustering, (3) identify genes that are informative
about the clusters.
</p>

<p>
The robustness of downstream results to filtering choices is not
systematically studied. However, one obvious area that does not appear to be
adequately explored is using existing statistical methods for black box
function optimization (e.g.,
<a href="http://papers.nips.cc/paper/4522-practical-bayesian-optimization">Snoek et
al. 2012</a>, <a href="https://dl.acm.org/doi/abs/10.5555/3122009.3242042">Li et
al. 2017</a>) in this setting, using metrics such as the
<a href="https://en.wikipedia.org/wiki/Rand_index#Adjusted_Rand_index">Adjusted Rand
index</a> of the cluster labels, the hold-out prediction accuracy of cluster
labels from gene expression, or the correlation of cluster &ldquo;centroids&rdquo; to
known/previously identified cell types as the optimization criterion. (It is
possible that some groups have tried these approaches and found they did not
work well.)
</p>

<p>
Commonly used representations such as PCA have statistical models underlying
them (e.g., <a href="https://dx.doi.org/10.1111/1467-9868.00196">Tipping and Bishop
1999</a>), which could in principle be used to produce uncertainty in the
representation. It is clear the propagating that uncertainty of certain
methods is very difficult, especially when those methods do not have an
explicit statistical model underlying them. However, propagating uncertainty
is easy in generative models that simultaneously estimate low-dimensional
representations and high-dimensional parameters of interest (e.g.,
differential expression analysis in a variational auto-encoder, as in
<a href="https://dx.doi.org/10.1038/s41592-018-0229-2">Lopez et al. 2018</a>). The
drawback of such approaches is that they require either more complex
inference algorithms, require solving more difficult optimization problems,
or require more work to interpret the results.
</p>

<p>
Identifying genes that are informative about cluster labels while quantifying
uncertainty in the identified genes is a variable selection problem, which is
well-studied (e.g., <a href="https://dx.doi.org/10.1111/rssb.12388">Wang et
al. 2020</a>).
</p>

<p>
<b>Scaling to higher dimensionalities: more cells, more features, and broader
coverage.</b> There are two fundamental questions: (1) do the software
implementations of common models/methods run with reasonable computational
budgets on time/space, and (2) do the models/methods support multiple types
of measurements (either simultaneously on the same samples or not)?
</p>

<p>
The first issue requires concentrated effort in software engineering, as
opposed to effort in new experimental protocols, data generation, statistical
models, or likely even in inference algorithms. This area is one in which
there are few major players (<a href="https://satijalab.org/seurat/">Seurat</a>,
<a href="https://scanpy.readthedocs.io/en/stable/">scanpy</a>), but they have the
majority of resources, attention, adoption, and traction. The second issue
requires the development of novel, linked statistical models in the case of
simultaneous measurements, which in turn likely requires specialized software
implementations. Further, there may be fundamental statistical problems in
the case of &ldquo;integrating&rdquo; non-simultaneous measurements, in the form of
un-correctable biases due to systematic technical differences between
experiments.
</p>
</div>
</div>

<div id="outline-container-orgcf2a286" class="outline-2">
<h2 id="orgcf2a286">Challenges in single-cell transcriptomics</h2>
<div class="outline-text-2" id="text-orgcf2a286">
<p>
<b>Handling sparsity in single-cell RNA sequencing.</b> This discussion is moot
(<a href="https://dx.doi.org/10.1101/2020.04.07.030007">Sarkar and Stephens
2020</a>). &ldquo;Dropouts&rdquo; are not a problem, &ldquo;imputation&rdquo; does not make sense, and
sparsity is a problem only for (space and time) efficient, robust software
implementations. The only open research question here is whether denoising is
just an interesting side-effect of fitting generative models for statistical
tasks of true interest, or whether denoised estimates of true gene expression
can be substituted for the data matrix in a modular approach for those
tasks. There is some indication that for visualization, at least, there is
interest in the denoised gene expression levels (e.g.,
<a href="https://academic.oup.com/bioinformatics/advance-article-abstract/doi/10.1093/bioinformatics/btab003/6103785">Alquira-Hernandez
and Powell 2021</a>).
</p>

<p>
<b>Defining flexible statistical frameworks for discovering complex
differential patterns in gene expression.</b> In what sense are current DE
frameworks &ldquo;inflexible&rdquo;? The paper suggests two senses: (1) DE needs fixed
assignments of samples to conditions, and (2) DE only considers mean
shifts. The first issue matters only when the assignments come from some
clustering algorithm, which generally has an uncertainty that is not
computed. This issue could be easily resolved by a model-based clustering
algorithm that assigns a posterior probability of cluster assignment. (It
could also be replaced by a different problem, of assessing DE in a topic
model.) The second issue relates to the general problem of detecting whether
two distributions are different from each other in a statistically meaningful
way, which seems to be a much harder problem, and one which has an unclear
biological interpretation. There has been some work in linking the
distribution of true gene expression levels to genetic
(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4019916/">Yvert 2013</a>) and
environmental
(<a href="https://link.springer.com/article/10.1186/s13059-016-1077-y">Korthauer et
al. 2016</a>) perturbations. However, there are still
<a href="diff-var.html#poisson-beta">fundamental disconnects</a> between these
ideas and e.g., theory of the kinetics of transcriptional regulation.
</p>

<p>
<b>Mapping single cells to a reference atlas.</b> The problems in this section
boil down to the statistical problem of semi-supervised learning of cell
clusters in new data sets, using reference data as labeled auxiliary
data. This problem seems to run into a fundamental limitation, that there
will be systematic differences in measurement error between new data sets and
reference data sets. This limitation suggests that what is needed is not just
labeled reference data, but a full generative model for the reference data
which can be updated with the new data set. Existing methods have already
begun to approach this idea
(<a href="https://www.nature.com/articles/s41592-019-0537-1">Wang et al. 2019</a>,
<a href="https://www.biorxiv.org/content/10.1101/2020.07.16.205997v1/">Lotfollahi et
al. 2020</a>).
</p>

<p>
On the one hand, distributing a full generative model could be easier than
distributing a high-dimensional cell atlas (c.f.,
<a href="https://pytorch.org/docs/stable/torchvision/models.html">pre-trained
Inception v3</a>). On the other hand, it is unclear how to make this sort of
framework widely accessible to non-experts, or even extensible by other
expert researchers.
</p>

<p>
<b>Generalizing trajectory inference.</b> What is &ldquo;trajectory inference&rdquo;? The
paper suggests a model in which cells undergoing some biological process move
in a state space parameterized by transcriptome, proteome, epigenome,
etc. Then, what would it mean to &ldquo;generalize&rdquo; trajectory inference? The paper
never addresses this question, but instead discusses the problems of: (1)
performing trajectory inference from different types of measurements, (2)
performing (parallel) trajectory inference from different initial conditions,
and (3) making rigorous statistical comparisons between estimated
trajectories.
</p>

<p>
One goal of trajectory inference is to learn about critical points,
characterized by specific changes in e.g., transcriptomes, in biological
processes of interest. Such critical points might correspond to bifurcations
of a relatively homogeneous population of cells responding to some stimulus
(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5365145/">LÃ¶nnberg et
al. 2017</a>), or more general branching from an initial pluripotent state
(<a href="https://science.sciencemag.org/content/360/6392/eaar3131.abstract">Farrell
et al. 2017</a>).
</p>

<p>
It is unclear that trajectory inference from single cell epigenomic or
proteomic data actually requires new statistical methods. However, this is in
part because the generative models for these processes (in particular, the
measurement error) have not been rigorously laid out. It is clear that
trajectory inference from simultaneous measurements of transcriptome,
proteome, etc. in each cell (e.g. CITE-Seq,
<a href="https://www.nature.com/articles/nmeth.4380">Stoeckius et al. 2017</a>) is
principled, and would require (relatively) simple development of new, linked
statistical models (in which different measurement models depend on a common
latent variable model). The situation of integrating independent trajectories
learned from transcriptome, etc. seems much more difficult (due to systematic
differences between measurements), and potentially unsolvable.
</p>

<p>
One glaring omission in this discussion is the use of known time points of
collection as prior information in trajectory inference. In general, cells do
not proceed lock-step through responses to stimuli, and the use of multiple
collections seems like the obvious choice to characterize temporal processes
at single cell resolution. However, one could argue that such experimental
designs are still unlikely to discover new biology, since the time points for
collections are typically chosen based on prior knowledge. However, the
statistical problems here seem much more tractable.
</p>

<p>
<b>Finding patterns in spatially resolved measurements.</b> The paper indicates
that the fundamental question is assigning samples to cell types/compartments
given gene expression and spatial coordinates. Why would this be an
interesting thing to do? In the case of slide-based approaches, one issue is
that measurement are not made at single-cell resolution, but at few-cell
resolution, and so (unwanted) spatial variation in cell type abundance may
lead to false positive discoveries of spatial patterns in gene expression
(e.g., <a href="https://www.nature.com/articles/s41587-021-00830-w">Cable et
al. 2020</a>,
<a href="https://www.biorxiv.org/content/10.1101/2020.11.15.378125v1">Kleschchevnikov
et al. 2021</a>).
</p>

<p>
<a href="spatial-vae.html">My own list of fundamental questions</a> in spatial
measurements is very different: (1) what are the predominant patterns of
spatial gene expression, (2) which spatial structures define (or are defined
by) those patterns, and (3) which genes co-vary spatially? In my view,
answering these questions are the first step towards understanding how
(perturbations to) cell-level (molecular) phenotypes lead to (changes in)
tissue-level phenotypes. I argue that understanding how cell-level phenotypes
lead to tissue-level phenotypes is the next major hurdle in connecting
disease-associated genetic variation to disease susceptibility/outcomes.
</p>

<p>
The paper also indicates that there are is also a problem of &ldquo;resolution&rdquo; in
spatial coherence of cells. My idea to use convolutional neural nets to
parameterize spatial expression models naturally handles this case, since
stacked convolutions naturally encode spatial coherence at different scales.
</p>
</div>
</div>

<div id="outline-container-org2bf77d5" class="outline-2">
<h2 id="org2bf77d5">Challenges in single-cell genomics</h2>
<div class="outline-text-2" id="text-org2bf77d5">
<p>
<b>Dealing with errors and missing data in the identification of variation from
single-cell DNA sequencing data.</b>
</p>
</div>
</div>

<div id="outline-container-orge9a2016" class="outline-2">
<h2 id="orge9a2016">Challenges in single-cell phylogenomics</h2>
<div class="outline-text-2" id="text-orge9a2016">
<p>
<b>Scaling phylogenetic models to many cells and many sites.</b>
</p>

<p>
<b>Integrating multiple types of variation into phylogenetic models.</b>
</p>

<p>
<b>Inferring population genetic parameters of tumor heterogeneity by model
integration.</b>
</p>
</div>
</div>

<div id="outline-container-org7c4f07b" class="outline-2">
<h2 id="org7c4f07b">Overarching challenges</h2>
<div class="outline-text-2" id="text-org7c4f07b">
<p>
<b>Integration of single-cell data across samples, experiments, and types of
measurement.</b>
</p>

<p>
<b>Validating and benchmarking analysis tools for single-cell measurements.</b>
</p>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2021-03-25 Thu 18:48</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
