<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-05-08 Fri 01:05 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Single cell QTL mapping via sparse multiple regression</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Single cell QTL mapping via sparse multiple regression</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgd860ef5">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org4ae132f">Results</a>
<ul>
<li><a href="#org6155286">Fine mapping dispersion effects</a></li>
<li><a href="#org8e84eda"><i>SKP1</i></a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orgd860ef5" class="outline-2">
<h2 id="orgd860ef5">Introduction</h2>
<div class="outline-text-2" id="text-orgd860ef5">
<p>
<a href="https://dx.doi.org/10.1371/journal.pgen.1008045">Sarkar et al. 2019</a>
discovered mean and variance effect QTLs using a modular approach, first
fitting the model \( 
  \DeclareMathOperator\Gam{Gamma}
  \DeclareMathOperator\N{\mathcal{N}}
  \DeclareMathOperator\Poi{Poisson}
  \DeclareMathOperator\argmin{arg min}
  \newcommand\mf{\mathbf{F}}
  \newcommand\mg{\mathbf{G}}
  \newcommand\ml{\mathbf{L}}
  \newcommand\mx{\mathbf{X}}
  \newcommand\vb{\mathbf{b}}
  \newcommand\vl{\mathbf{l}}
  \newcommand\vx{\mathbf{x}}
  \)
</p>

\begin{align}
  x_{ij} \mid x_{i+}, \lambda_{ij} &\sim \Poi(x_{i+} \lambda_{ij})\\
  \lambda_{ij} \mid \mu_{ij}, \phi_{ij}, \pi_{ij} &\sim \pi_{ij} \delta_0(\cdot) + (1 - \pi_{ij}) \Gam(\phi_{ij}^{-1}, \mu_{ij}^{-1} \phi_{ij}^{-1})\\
  \ln \mu_{ij} &= (\ml \mf_\mu')_{ij}\\
  \ln \phi_{ij} &= (\ml \mf_\phi')_{ij}\\
  \operatorname{logit} \pi_{ij} &= (\ml \mf_\pi')_{ij},
\end{align}

<p>
where 
</p>

<ul class="org-ul">
<li>\(x_{ij}\) is the number of molecules of gene \(j = 1, \ldots, p\) observed
in cell \(i = 1, \ldots, n\)</li>
<li>\(x_{i+} \triangleq \sum_j x_{ij}\) is the total number of molecules
observed in sample \(i\)</li>
<li>cells are taken from \(m\) donor individuals, \(\ml\) is \(n \times m\),
and each \(\mf_{(\cdot)}\) is \(p \times m\)</li>
<li>assignments of cells to donors (loadings) \(l_{ik} \in \{0, 1\}, k = 1,
    \ldots, m\) are known and fixed.</li>
</ul>

<p>
and then using a standard QTL mapping approach (e.g.,
<a href="https://dx.doi.org/10.1038/nature10808">Degner et al. 2012</a>,
<a href="https://dx.doi.org/10.1126/science.1242429">McVicker et al. 2013</a>) to
discover genetic effects on gene expression mean \(E[\lambda_{ij}]\),
dispersion \(\phi_{ij}\), and variance \(V[\lambda_{ij}]\). Critically, these
quantities describe the distribution of latent gene expression, removing the
effect of measurement noise
(<a href="https://dx.doi.org/10.1101/2020.04.07.030007">Sarkar and Stephens 2020</a>).
</p>

<p>
This approach lost substantial power to detect effect on mean gene expression
compared to bulk RNA-seq on the same (number of) samples
(<a href="https://dx.doi.org/10.1101/gr.224436.117">Banovich et al. 2018</a>). One
approach which might increase power would be to incorporate genotype into the
expression model for gene \(j\)
</p>

\begin{align}
  \lambda_{i} &= \Gam(\phi_{i}^{-1}, \mu_{i}^{-1} \phi_{i}^{-1})\\
  \ln\mu_{i} &= \ml(\mg\vb_{\mu} + b_{\mu 0})\\
  \ln\phi_{i} &= \ml(\mg\vb_{\phi} + b_{\phi 0}),
\end{align}

<p>
where
</p>

<ul class="org-ul">
<li>\(\mg\) denotes genotype (\(m \times s\))</li>
<li>\(\vb_{\mu}\), \(\vb_{\phi}\) denote genetic effects (\(s \times 1)\)</li>
<li>\(b_{\mu 0}\), \(b_{\phi 0}\) denote (scalar) intercepts</li>
</ul>

<p>
This approach would increase power by pooling samples across individuals with
same (similar) genotypes, and is similar to previously proposed approaches
(e.g., <a href="https://www.genetics.org/content/188/2/435">Rönnegård and Valdar
2011</a>,
<a href="https://bmcgenet.biomedcentral.com/articles/10.1186/1471-2156-13-63">Rönnegård
and Valdar 2012</a>). To fit this model for \(s > m\), we can further assume a
sparsity-inducing prior
</p>

\begin{align}
  \vb_{\mu} &\sim \pi_{\mu} \delta_0(\cdot) + (1 - \pi_{\mu}) \N(0, \sigma_{\mu}^2)\\
  \vb_{\phi} &\sim \pi_{\phi} \delta_0(\cdot) + (1 - \pi_{\phi}) \N(0, \sigma_{\phi}^2).
\end{align}

<p>
The prior is non-conjugate to the likelihood, so we perform variational
inference (Blei et al. 2017), optimizing a stochastic objective
(<a href="https://arxiv.org/abs/1312.6114">Kingma and Welling 2014</a>,
<a href="https://arxiv.org/abs/1401.4082">Rezende et al. 2014</a>,
<a href="http://proceedings.mlr.press/v32/titsias14.html">Titsias and
Lázaro-Gredilla 2014</a>,
<a href="http://proceedings.mlr.press/v33/ranganath14.html">Ranganath et al. 2014</a>)
via gradient descent. We use a local reparameterization
(<a href="https://papers.nips.cc/paper/5666-variational-dropout-and-the-local-reparameterization-trick">Kingma
et al. 2015</a>) to speed up convergence
(<a href="https://dx.doi.org/10.1101/107623">Park et al. 2017</a>).
</p>
</div>
</div>

<div id="outline-container-org0620222" class="outline-2">
<h2 id="setup"><a id="org0620222"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> tabix
<span class="org-keyword">import</span> torch
<span class="org-keyword">import</span> torch.utils.data <span class="org-keyword">as</span> td
<span class="org-keyword">import</span> txpred
<span class="org-keyword">import</span> txpred.models.susie
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org4ae132f" class="outline-2">
<h2 id="org4ae132f">Results</h2>
<div class="outline-text-2" id="text-org4ae132f">
</div>
<div id="outline-container-org6155286" class="outline-3">
<h3 id="org6155286">Fine mapping dispersion effects</h3>
<div class="outline-text-3" id="text-org6155286">
<p>
The motivation for restricting QTL mapping to 100kb windows was to reduce
the number of single SNP tests, and therefore increase the power to detect
associations at a fixed FDR (intuitively, the BH procedure penalizes each
test based on the total number of tests). Using fast multiple regression
methods with Bayesian variable selection guarantees, we can simply avoid
this issue. Read the estimated dispersion parameters.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">log_phi</span> = pd.read_csv(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-qtl/data/density-estimation/design1/zi2-log-phi.txt.gz'</span>, index_col=0, sep=<span class="org-string">' '</span>).T
log_phi.shape
</pre>
</div>

<pre class="example">
(54, 9957)

</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">y</span> = log_phi[<span class="org-string">'ENSG00000113558'</span>]
<span class="org-variable-name">dat</span> = pd.read_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/skp1.vcf.gz'</span>, skiprows=2, sep=<span class="org-string">'\t'</span>)
<span class="org-variable-name">x</span> = dat.<span class="org-builtin">filter</span>(like=<span class="org-string">'NA'</span>, axis=1).T
<span class="org-variable-name">x.columns</span> = dat[<span class="org-string">'POS'</span>]
<span class="org-variable-name">x</span> = x.sub(x.mean(axis=1), axis=0)
<span class="org-variable-name">x</span> = x.div(x.std(axis=1), axis=0)
<span class="org-variable-name">x</span>, <span class="org-variable-name">y</span> = x.align(y, axis=0, join=<span class="org-string">'inner'</span>)
x.shape
</pre>
</div>

<pre class="example">
(54, 5778)

</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
rpy2.robjects.pandas2ri.activate()
<span class="org-variable-name">susie</span> = rpy2.robjects.packages.importr(<span class="org-string">'susieR'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">fit</span> = susie.susie(x.values, y.values, L=10)
susie.susie_get_cs(fit, x.values).rx2(<span class="org-string">'cs'</span>)
</pre>
</div>

<pre class="example">
&lt;rpy2.rinterface.NULLType object at 0x7fb121e3e410&gt; [RTYPES.NILSXP]

</pre>
</div>
</div>

<div id="outline-container-org8e84eda" class="outline-3">
<h3 id="org8e84eda"><i>SKP1</i></h3>
<div class="outline-text-3" id="text-org8e84eda">
<p>
Read the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">y</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/ipsc/ipsc.h5ad'</span>)
</pre>
</div>

<p>
<i>SKP1</i> was the top eQTL discovered in Sarkar et al. 2019. Extract the
genotypes around the TSS.
</p>

<div class="org-src-container">
<pre class="src src-ipython">y.var[y.var[<span class="org-string">'name'</span>] == <span class="org-string">'SKP1'</span>]
</pre>
</div>

<pre class="example">
chr      start        end  name strand      source
index
ENSG00000113558  hs5  133484633  133512729  SKP1      -  H. sapiens
</pre>

<div class="org-src-container">
<pre class="src src-sh">tabix -h /project2/mstephens/aksarkar/projects/singlecell-qtl/data/scqtl-mapping/yri-120-dosages.vcf.gz chr5:$((133484633-1000000))-$((133484633+1000000)) | gzip &gt;skp1.vcf.gz
</pre>
</div>

<p>
Read the genotypes, and align it to the individuals with single cell
measurements.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = pd.read_csv(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/skp1.vcf.gz'</span>, skiprows=2, sep=<span class="org-string">'\t'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">L</span> = pd.get_dummies(y.obs[<span class="org-string">'chip_id'</span>])
<span class="org-variable-name">x</span> = dat.<span class="org-builtin">filter</span>(like=<span class="org-string">'NA'</span>, axis=1)
<span class="org-variable-name">x.index</span> = dat[<span class="org-string">'POS'</span>]
<span class="org-variable-name">x</span> = x.sub(x.mean(axis=1), axis=0)
<span class="org-variable-name">x</span> = x.div(x.std(axis=1), axis=0)
<span class="org-variable-name">L</span>, <span class="org-variable-name">x</span> = L.align(x, axis=1, join=<span class="org-string">'inner'</span>)
x.T.shape
</pre>
</div>

<pre class="example">
(54, 5778)

</pre>

<p>
Fit the sparse regression model.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">data</span> = td.DataLoader(
  td.TensorDataset(
    torch.tensor(L.values, dtype=torch.<span class="org-builtin">float</span>),
    torch.tensor(y[:,y.var[<span class="org-string">'name'</span>] == <span class="org-string">'SKP1'</span>].X.A, dtype=torch.<span class="org-builtin">float</span>),
    torch.tensor(y.obs[<span class="org-string">'mol_hs'</span>].values, dtype=torch.<span class="org-builtin">float</span>)),
  batch_size=128, shuffle=<span class="org-constant">True</span>)
<span class="org-variable-name">fit</span> = (txpred.models.susie.PoisGamRegression(m=L.shape[1], p=x.shape[0])
       .fit(torch.tensor(x.values.T, dtype=torch.<span class="org-builtin">float</span>), data, n_epochs=80, trace=<span class="org-constant">False</span>))
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(4, 3)
<span class="org-variable-name">grid</span> = np.array(x.index) / 1e6
<span class="org-variable-name">pip</span> = torch.sigmoid(fit.mean_prior.logits.detach()).numpy().ravel()
ax[0].scatter(grid[pip &gt; 1e-2], pip[pip &gt; 1e-2], s=2, c=<span class="org-string">'k'</span>)
<span class="org-keyword">for</span> t <span class="org-keyword">in</span> grid[pip &gt; .5]:
  <span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
    a.axvline(x=t, lw=.5, c=<span class="org-string">'0.7'</span>, zorder=-1)
ax[0].set_ylabel(r<span class="org-string">'PIP $\ln(\mu)$'</span>)
ax[0].set_xlim(grid.<span class="org-builtin">min</span>(), grid.<span class="org-builtin">max</span>())
ax[0].set_ylim(0, 1)
<span class="org-variable-name">pip</span> = torch.sigmoid(fit.inv_disp_prior.logits.detach()).numpy().ravel()
ax[1].scatter(grid[pip &gt; 1e-2], pip[pip &gt; 1e-2], s=2, c=<span class="org-string">'k'</span>)
<span class="org-keyword">for</span> t <span class="org-keyword">in</span> grid[pip &gt; .5]:
  <span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
    a.axvline(x=t, lw=.5, c=<span class="org-string">'0.7'</span>, zorder=-1)
ax[1].set_xlabel(<span class="org-string">'Position (MB)'</span>)
ax[1].set_ylabel(r<span class="org-string">'PIP $\ln(\phi)$'</span>)
<span class="org-variable-name">t</span> = y.var.loc[y.var[<span class="org-string">'name'</span>] == <span class="org-string">'SKP1'</span>, <span class="org-string">'start'</span>][0] / 1e6
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax:
  a.axvline(x=t + .1, c=<span class="org-string">'b'</span>, lw=.5, zorder=3)
  a.axvline(x=t - .1, c=<span class="org-string">'b'</span>, lw=.5, zorder=3)
ax[0].set_title(<span class="org-string">'SKP1'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/scqtl.org/skp1.png" alt="skp1.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-05-08 Fri 01:05</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
