<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2021-05-30 Sun 16:20 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>GOM model for spatial transcriptomic data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">GOM model for spatial transcriptomic data</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org2e779ea">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org5b9c931">Results</a>
<ul>
<li><a href="#org6037e95">Visium</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org2e779ea" class="outline-2">
<h2 id="org2e779ea">Introduction</h2>
<div class="outline-text-2" id="text-org2e779ea">
<p>
<a href="https://pubmed.ncbi.nlm.nih.gov/10548103/">Lee and Seung 1999</a> described
how (Poisson) non-negative matrix factorization (NMF) can be used to learn
&ldquo;parts&rdquo; of faces/semantic concepts that combine additively to generate
observed images/text. Here, we investigate the use of NMF to learn spatial
structures from <i>spatial transcriptomic</i> data
(<a href="https://science.sciencemag.org/content/353/6294/78">Ståhl et al. 2016</a>,
<a href="https://www.nature.com/articles/s41592-019-0548-y">Vikovic et al. 2019</a>,
<a href="https://science.sciencemag.org/content/353/6294/78">Rodriques et
al. 2019</a>).
</p>
</div>
</div>

<div id="outline-container-org5489025" class="outline-2">
<h2 id="setup"><a id="org5489025"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> pickle
<span class="org-keyword">import</span> scanpy <span class="org-keyword">as</span> sc
<span class="org-keyword">import</span> scipy.sparse <span class="org-keyword">as</span> ss
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> scmodes
<span class="org-keyword">import</span> sklearn.datasets <span class="org-keyword">as</span> skd
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
rpy2.robjects.pandas2ri.activate()
<span class="org-variable-name">matrix</span> = rpy2.robjects.packages.importr(<span class="org-string">'Matrix'</span>)
<span class="org-variable-name">fasttopics</span> = rpy2.robjects.packages.importr(<span class="org-string">'fastTopics'</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org5b9c931" class="outline-2">
<h2 id="org5b9c931">Results</h2>
<div class="outline-text-2" id="text-org5b9c931">
</div>
<div id="outline-container-org6037e95" class="outline-3">
<h3 id="org6037e95">Visium</h3>
<div class="outline-text-3" id="text-org6037e95">
<p>
Read the <a href="spatial-vae.html#orgde57de0">pre-processed data</a>. Filter out
genes with no observed data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = anndata.read_h5ad(<span class="org-string">'/project2/mstephens/aksarkar/projects/singlecell-ideas/data/spatial/V1_Breast_Cancer_Block_A_Section_1/dat.h5ad'</span>)
sc.pp.filter_genes(dat, min_cells=1)
dat
</pre>
</div>

<pre class="example">
AnnData object with n_obs × n_vars = 3798 × 24923
obs: 'barcode', 'in_tissue', 'row', 'col', 'pxl_row', 'pxl_col'
var: 'gene', 'name', 'featuretype', 'n_cells'
</pre>

<p>
Fit a topic model, ignoring spatial information.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = dat.X.tocoo()
<span class="org-variable-name">y</span> = matrix.sparseMatrix(i=pd.Series(temp.row + 1), j=pd.Series(temp.col + 1), x=pd.Series(temp.data), dims=pd.Series(temp.shape))
<span class="org-variable-name">res</span> = fasttopics.fit_poisson_nmf(y, k=10, numiter=100, method=<span class="org-string">'scd'</span>, control=rpy2.robjects.ListVector({<span class="org-string">'extrapolate'</span>: <span class="org-constant">True</span>}), verbose=<span class="org-constant">True</span>)
<span class="org-variable-name">res</span> = fasttopics.poisson2multinom(res)
</pre>
</div>

<p>
Write out the fitted model.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">with</span> <span class="org-builtin">open</span>(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/visium-k-10-topic-model.pkl'</span>, <span class="org-string">'wb'</span>) <span class="org-keyword">as</span> f:
  pickle.dump(res, f)
</pre>
</div>

<p>
Look at the loadings on each topic, spatially organized.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 6, sharex=<span class="org-constant">True</span>, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(7, 3)
<span class="org-keyword">for</span> k, a <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(ax[:,:5].ravel()):
  <span class="org-variable-name">temp</span> = ss.coo_matrix((res.rx2(<span class="org-string">'L'</span>)[:,k], (dat.obs[<span class="org-string">'row'</span>], dat.obs[<span class="org-string">'col'</span>] // 2)), shape=(80, 64))
  a.imshow(temp.A, cmap=colorcet.cm[<span class="org-string">'fire'</span>])
  a.set_title(f<span class="org-string">'Topic {k + 1}'</span>)
  a.set_xticks([])
  a.set_yticks([])
<span class="org-keyword">for</span> a <span class="org-keyword">in</span> ax[:,5]:
  a.axis(<span class="org-string">'off'</span>)
<span class="org-variable-name">cb</span> = fig.colorbar(matplotlib.cm.ScalarMappable(norm=matplotlib.colors.Normalize(vmin=0, vmax=1),
                                               cmap=colorcet.cm[<span class="org-string">'fire'</span>]),
                  ax=ax[:,5].ravel(),
                  anchor=(0, 0.5),
                  shrink=0.75)
cb.set_label(<span class="org-string">'Loading'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/spatial-gom.org/visium-k-10.png" alt="visium-k-10.png">
</p>
</div>

<p>
Look at the top (highest probability) genes in each topic.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">freq_genes</span> = pd.DataFrame({k: dat.var.iloc[np.argsort(-res.rx2(<span class="org-string">'F'</span>)[:,k])[:10]][<span class="org-string">'name'</span>].values <span class="org-keyword">for</span> k <span class="org-keyword">in</span> <span class="org-builtin">range</span>(10)})
freq_genes
</pre>
</div>

<pre class="example">
0        1       2       3        4        5       6       7       8  \
0    COX6C    RPL41    CD74   COX6C   MT-CO1   IGFBP5    IGKC  MT-CO3     MGP
1   MT-CO1     KRT8     FTL   KRT19   MT-CO3     KRT8   IGHG3     FTL  IGFBP5
2    RPL41    RPLP1     B2M   RPL41   MT-CO2    RPL41   IGHG4  COL1A1   RPL41
3     KRT8    RPS27  TMSB10  IGFBP5  MT-ATP6    COX6C   IGLC2   KRT19   KRT19
4    KRT19    KRT18   RPL41    CPB1   MT-ND4    RPS27   IGHG1  COL1A2   COX6C
5   MT-ND1    KRT19    APOE  TMSB10   MT-ND3    KRT18   IGHA1   RPLP1   GAPDH
6    SERF2  S100A11  TMSB4X   ACTG1   MT-ND1    RPLP1   IGLC1    KRT8   RPL30
7  SLC39A6    RPL13   RPS27   RPL13   MT-CYB    KRT19    IGHM     MIF    TFF3
8   MT-CO3     MUC1  EEF1A1   RPLP1   MT-ND2     FTH1   IGLC3   SERF2  MT-CO1
9    ACTG1     FTH1    ACTB    KRT8    RPL41  S100A11  JCHAIN     FN1   RPS18

9
0  COX6C
1    B2M
2  HLA-B
3  HLA-A
4  IFI27
5  KRT19
6   IFI6
7   CD74
8   FTH1
9  HLA-C
</pre>

<p>
Look at distinctive genes in each topic, i.e., genes with maximum minimum KL
divergence from all other topics.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">top_genes_idx</span>(F, k, num_genes=10):
  <span class="org-variable-name">fk</span> = F[:,k].reshape(-1, 1)
  <span class="org-variable-name">temp</span> = fk * (np.log(fk) - np.log(F)) - fk + F
  <span class="org-keyword">return</span> np.argsort(-temp[:,np.arange(temp.shape[1]) != k].<span class="org-builtin">min</span>(axis=1))[:num_genes]

<span class="org-variable-name">top_genes</span> = pd.DataFrame({k: dat.var.iloc[top_genes_idx(np.array(res.rx2(<span class="org-string">'F'</span>)), k)][<span class="org-string">'name'</span>].values <span class="org-keyword">for</span> k <span class="org-keyword">in</span> <span class="org-builtin">range</span>(10)})
top_genes
</pre>
</div>

<pre class="example">
0       1         2        3        4         5      6       7  \
0    COX6C  CXCL14    TMSB4X     CPB1   MT-CO1    CRISP3   IGKC  COL1A1
1  SLC39A6   CCND1     KRT19    COX6C   MT-CO2   SLITRK6  IGHG3  COL1A2
2   MT-ND1    AGR2      CD74   FCGR3B   MT-CO3  C6orf141  IGLC2   SPARC
3   MT-CO1  TTLL12    IGFBP5      CFB  MT-ATP6    SERHL2  IGHG4     FN1
4    RPS15   GFRA1     HLA-B  TNFSF10   MT-ND3     H3F3A  IGHG1  COL3A1
5     DDR1   DEGS1   HLA-DRA     TFF3   MT-ND4    IGFBP5  IGHA1  LGALS1
6      SCD  MT-ND1      KRT8    KRT19   MT-CYB      CD74  IGLC3  MT-CO3
7    RPL17  TCEAL4       LYZ  COLEC12   MT-ND1     VTCN1  IGLC1   POSTN
8     SNCG   ARMT1     KRT18  SLC39A6   MT-ND2     NUPR1  IGHG2  COL6A2
9    RPL13    GNG5  HLA-DPB1    REPS2   MT-ND5   S100A13   IGHM   TAGLN

8      9
0         MGP  HLA-B
1         ALB   IFI6
2       S100G  COX6C
3    SERPINA3  IFI27
4  ZNF350-AS1  HLA-C
5       GAPDH  HLA-A
6        TFF3  ISG15
7      S100A6  RPL13
8        STC2   CD74
9        CD74  RPS18
</pre>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = res.rx2(<span class="org-string">'F'</span>)[dat.var[<span class="org-string">'name'</span>].isin(<span class="org-builtin">set</span>(freq_genes.values.ravel()))]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(7.5, 3)
<span class="org-variable-name">im</span> = plt.imshow(temp.T, cmap=colorcet.cm[<span class="org-string">'bmy'</span>], norm=matplotlib.colors.LogNorm(vmin=temp.<span class="org-builtin">min</span>(), vmax=temp.<span class="org-builtin">max</span>()))
plt.xlabel(<span class="org-string">'Gene'</span>)
plt.ylabel(<span class="org-string">'Topic'</span>)
plt.xticks(np.arange(temp.shape[0]), dat.var.loc[dat.var[<span class="org-string">'name'</span>].isin(<span class="org-builtin">set</span>(freq_genes.values.ravel())), <span class="org-string">'name'</span>].values, size=8, rotation=90)
plt.yticks(np.arange(temp.shape[1]), 1 + np.arange(temp.shape[1]), size=8)
<span class="org-variable-name">cb</span> = plt.colorbar(im, shrink=0.5)
cb.set_label(<span class="org-string">'Relative gene expression'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/spatial-gom.org/visium-k-10-genes.png" alt="visium-k-10-genes.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2021-05-30 Sun 16:20</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
