<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2020-10-01 Thu 17:04 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Convolutional VAE for spatial transcriptomic data</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Convolutional VAE for spatial transcriptomic data</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org396fcc7">Introduction</a></li>
<li><a href="#setup">Setup</a></li>
<li><a href="#org2363802">Methods</a>
<ul>
<li><a href="#org0f8899d">Data</a></li>
</ul>
</li>
<li><a href="#org054560c">Results</a>
<ul>
<li><a href="#orge61c045">MNIST</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org396fcc7" class="outline-2">
<h2 id="org396fcc7">Introduction</h2>
<div class="outline-text-2" id="text-org396fcc7">
<p>
There is great interest in analyzing transcriptomic profiles in complex
tissues measured at single (or few) cell resolution jointly with associated
spatial information
(<a href="https://science.sciencemag.org/content/353/6294/78">Ståhl et al. 2016</a>,
<a href="https://www.nature.com/articles/s41592-019-0548-y">Vikovic et al. 2019</a>,
<a href="https://science.sciencemag.org/content/353/6294/78">Rodriques et
al. 2019</a>). Such analysis could elucidate the molecular basis for the
spatial organization of cell types/states
(<a href="https://dx.doi.org/10.1016/j.copbio.2017.02.004">Moor and Itzkovitz
2017</a>, <a href="https://science.sciencemag.org/content/358/6359/64">Lein et
al. 2017</a>), and could be useful for genomics-based histology. Existing
statistical methods for analyzing spatial transcriptomic data are largely
based on Gaussian processes
(<a href="https://dx.doi.org/10.1038/nmeth.4636">Svensson et al. 2018</a>,
<a href="https://www.nature.com/articles/nmeth.4634">Edsgärd et al. 2018</a>,
<a href="https://doi.org/10.1038/s41592-019-0701-7">Sun et al. 2020</a>). These
methods have proven to be successful in identifying genes that exhibit
statistically significant spatial patterns of gene expression (compared to a
null model of no structured spatial variation). However, these methods
analyze one gene at a time, and therefore lose information on gene
co-expression within and between samples. Further, they rely on a statistical
test to prioritize genes that could be relevant to observed spatial
structures (e.g., in associated imaging data).
</p>

<p>
For spatial transcriptomic data that are collected on a regular grid, we can
instead use convolutional neural networks (CNNs; Denker et al. 1989, LeCun et
al. 1989) to learn spatial structure. There are several advantages to the
convolutional approach. 
</p>

<ol class="org-ol">
<li>A single model can be learned on all genes, and therefore can exploit
covariance of gene expression induced by the gene regulatory network.</li>
<li>The model can learn a lower-dimensional representation of spatial
expression patterns, which can be used to identify genes with highly
similar spatial expression patterns</li>
<li>With associated imaging data, the convolutional filters can be jointly
learned between image features and gene expression features, which can
more directly associate cell type/state differences at the expression
level with e.g., differences in morphology.</li>
</ol>

<p>
The key idea of our approach is to combine a Poisson measurement model
(<a href="https://dx.doi.org/10.1101/2020.04.07.030007">Sarkar and Stephens 2020</a>),
</p>

\begin{equation}
  \newcommand\Pois{\operatorname{Pois}}
  \newcommand\N{\mathcal{N}}
  \newcommand\vx{\mathbf{x}}
  \newcommand\vz{\mathbf{z}}
  x_{ij} \mid s_i, \lambda_{ij} \sim \Pois(s_i \lambda_{ij})
\end{equation}

<p>
with a <i>spatial expression model</i>,
</p>

\begin{equation}
  \lambda_{ij} \mid \vz_j = (h(\vz_j))_i,
\end{equation}

<p>
where \(h\) denotes a CNN. To perform approximate Bayesian inference (on
\(\vz_j\)), we treat \(h\) as a decoder network in a variational autoencoder
(Kingma and Welling 2014) and introduce an encoder network
</p>

\begin{equation}
  \vz_j \mid \vx_{\cdot j} \sim \N(\mu(\vx_{\cdot j}), \sigma^2(\vx_{\cdot j})),
\end{equation}

<p>
where \(\mu, \sigma^2\) are also CNNs (Salimans et al. 2014). Unlike existing
application of VAEs to scRNA-seq data (e.g., Lopez et al. 2018), in the
spatial transcriptomics setting, inference is done over minibatches of genes
rather than samples.
</p>

<p>
The learned VAE can be used to answer several questions of interest:
</p>

<ol class="org-ol">
<li><i>What are the predominant patterns of spatial gene expression?</i> These are
learned as the low dimensional spatial representations of each gene
\(\vz_j\).</li>
<li><i>Which spatial structures define those patterns?</i> These can be recovered
using e.g., saliency maps to associate learned convolutional filters with
the specific latent variable \(\vz_j\) of interest.</li>
<li><i>Which genes co-vary spatially?</i> These can be recovered by clustering the
learned \(\vz_j\).</li>
</ol>
</div>
</div>

<div id="outline-container-orgb67d462" class="outline-2">
<h2 id="setup"><a id="orgb67d462"></a>Setup</h2>
<div class="outline-text-2" id="text-setup">
<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> torch
<span class="org-keyword">import</span> torch.utils.data <span class="org-keyword">as</span> td
<span class="org-keyword">import</span> torch.utils.tensorboard <span class="org-keyword">as</span> tb
<span class="org-keyword">import</span> torchvision <span class="org-keyword">as</span> tv
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-org2363802" class="outline-2">
<h2 id="org2363802">Methods</h2>
<div class="outline-text-2" id="text-org2363802">
</div>
<div id="outline-container-org0f8899d" class="outline-3">
<h3 id="org0f8899d">Data</h3>
<div class="outline-text-3" id="text-org0f8899d">
<ul class="org-ul">
<li><a href="https://support.10xgenomics.com/spatial-gene-expression/datasets/1.1.0/V1_Breast_Cancer_Block_A_Section_1">Breast
cancer</a> (10X Genomics Visium v1.1.0)</li>
<li><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE130682">Mouse
olfactory bulb</a> (HDST;
<a href="https://www.nature.com/articles/s41592-019-0548-y">Vikovic et
al. 2019</a>)</li>
<li>Human spinal cord
(<a href="https://science.sciencemag.org/content/364/6435/89.full">Maniatis et
al. 2019</a>)</li>
<li>Mouse cerebellum (SLIDE-Seq;
<a href="https://science.sciencemag.org/content/353/6294/78">Rodriques et
al. 2019</a>)</li>
<li>Human gingival tissue (ST;
<a href="https://www.nature.com/articles/s41598-018-27627-3">Lundmark et
al. 2018</a>)</li>
<li>Human prostate cancer (ST;
<a href="https://www.nature.com/articles/s41467-018-04724-5">Berglund et
al. 2018</a>)</li>
<li>Human melanoma (ST;
<a href="https://cancerres.aacrjournals.org/content/78/20/5970.short">Thrane et
al. 2018</a>)</li>
<li>Human heart (ST; <a href="https://www.nature.com/articles/s41598-018-27627-3">Asp
et al. 2017</a>)</li>
<li>Human breast cancer (ST;
<a href="https://science.sciencemag.org/content/353/6294/78">Ståhl et al. 2016</a>)</li>
<li>Mouse olfactory bulb
(ST; <a href="https://science.sciencemag.org/content/353/6294/78">Ståhl et al. 2016</a>)</li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org054560c" class="outline-2">
<h2 id="org054560c">Results</h2>
<div class="outline-text-2" id="text-org054560c">
</div>
<div id="outline-container-orge61c045" class="outline-3">
<h3 id="orge61c045">MNIST</h3>
<div class="outline-text-3" id="text-orge61c045">
<p>
Implement a VAE using convolutional layers instead of fully connected
layers.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">class</span> <span class="org-type">Encoder</span>(torch.nn.Module):
  <span class="org-keyword">def</span> <span class="org-function-name">__init__</span>(<span class="org-keyword">self</span>, input_h, input_w, output_dim):
    <span class="org-builtin">super</span>().__init__()
    <span class="org-keyword">self</span>.net = torch.nn.Sequential(
      torch.nn.Conv2d(in_channels=1, out_channels=32, kernel_size=3, padding=1),
      torch.nn.ReLU(),
      torch.nn.Conv2d(in_channels=32, out_channels=64, kernel_size=3, padding=1),
      torch.nn.ReLU(),
      torch.nn.Flatten(),
      torch.nn.Linear(64 * input_h * input_w, 1024),
      torch.nn.ReLU(),
      torch.nn.Linear(1024, 512),
      torch.nn.ReLU(),
    )
    <span class="org-keyword">self</span>.mean = torch.nn.Linear(512, output_dim)
    <span class="org-keyword">self</span>.scale = torch.nn.Sequential(
      torch.nn.Linear(512, output_dim),
      torch.nn.Softplus())

  <span class="org-keyword">def</span> <span class="org-function-name">forward</span>(<span class="org-keyword">self</span>, x):
    <span class="org-variable-name">h</span> = <span class="org-keyword">self</span>.net(x)
    <span class="org-keyword">return</span> <span class="org-keyword">self</span>.mean(h), <span class="org-keyword">self</span>.scale(h)

<span class="org-keyword">class</span> <span class="org-type">Decoder</span>(torch.nn.Module):
  <span class="org-keyword">def</span> <span class="org-function-name">__init__</span>(<span class="org-keyword">self</span>, input_dim, output_h, output_w):
    <span class="org-builtin">super</span>().__init__()
    <span class="org-keyword">self</span>.net1 = torch.nn.Sequential(
      torch.nn.Linear(input_dim, 512),
      torch.nn.ReLU(),
      torch.nn.Linear(512, 1024),
      torch.nn.ReLU(),
      torch.nn.Linear(1024, 64 * output_h * output_w),
      torch.nn.ReLU(),
    )
    <span class="org-keyword">self</span>.net2 = torch.nn.Sequential(
      torch.nn.ConvTranspose2d(in_channels=64, out_channels=32, kernel_size=3, padding=1),
      torch.nn.ReLU(),
      torch.nn.ConvTranspose2d(in_channels=32, out_channels=1, kernel_size=3, padding=1),
      torch.nn.Sigmoid(),
    )
    <span class="org-keyword">self</span>.output_h = output_h
    <span class="org-keyword">self</span>.output_w = output_w

  <span class="org-keyword">def</span> <span class="org-function-name">forward</span>(<span class="org-keyword">self</span>, x):
    <span class="org-variable-name">h</span> = <span class="org-keyword">self</span>.net1(x).reshape([-1, 64, <span class="org-keyword">self</span>.output_h, <span class="org-keyword">self</span>.output_w])
    <span class="org-keyword">return</span> <span class="org-keyword">self</span>.net2(h)

<span class="org-keyword">class</span> <span class="org-type">ConvVAE</span>(torch.nn.Module):
  <span class="org-keyword">def</span> <span class="org-function-name">__init__</span>(<span class="org-keyword">self</span>, input_h, input_w, latent_dim):
    <span class="org-builtin">super</span>().__init__()
    <span class="org-keyword">self</span>.encoder = Encoder(input_h, input_w, latent_dim)
    <span class="org-keyword">self</span>.decoder = Decoder(latent_dim, input_h, input_w)

  <span class="org-keyword">def</span> <span class="org-function-name">forward</span>(<span class="org-keyword">self</span>, x, n_samples, writer=<span class="org-constant">None</span>, global_step=<span class="org-constant">None</span>):
    <span class="org-variable-name">mean</span>, <span class="org-variable-name">scale</span> = <span class="org-keyword">self</span>.encoder(x)
    <span class="org-variable-name">kl</span> = torch.<span class="org-builtin">sum</span>(.5 * (1 - 2 * torch.log(scale) + (mean * mean + scale * scale)), dim=1)
    <span class="org-variable-name">qz</span> = torch.distributions.Normal(mean, scale).rsample(n_samples)
    <span class="org-variable-name">probs</span> = <span class="org-keyword">self</span>.decoder.forward(qz)
    <span class="org-variable-name">error</span> = torch.<span class="org-builtin">sum</span>(torch.mean(x * torch.log(probs) + (1 - x) * torch.log(1 - probs), dim=1), dim=[1, 2])
    <span class="org-keyword">if</span> writer <span class="org-keyword">is</span> <span class="org-keyword">not</span> <span class="org-constant">None</span>:
      writer.add_scalar(<span class="org-string">'loss/error'</span>, torch.<span class="org-builtin">sum</span>(error), global_step)
      writer.add_scalar(<span class="org-string">'loss/kl'</span>, torch.<span class="org-builtin">sum</span>(kl), global_step)
    <span class="org-variable-name">loss</span> = -torch.<span class="org-builtin">sum</span>(error - kl)
    <span class="org-keyword">if</span> torch.isnan(loss):
      <span class="org-keyword">raise</span> <span class="org-type">RuntimeError</span>
    <span class="org-keyword">return</span> loss

  <span class="org-keyword">def</span> <span class="org-function-name">fit</span>(<span class="org-keyword">self</span>, data, n_epochs=5, n_samples=<span class="org-constant">None</span>, log_dir=<span class="org-constant">None</span>, **kwargs):
    <span class="org-keyword">if</span> log_dir <span class="org-keyword">is</span> <span class="org-constant">None</span>:
      <span class="org-variable-name">writer</span> = <span class="org-constant">None</span>
    <span class="org-keyword">else</span>:
      <span class="org-variable-name">writer</span> = tb.SummaryWriter(log_dir)
    <span class="org-keyword">if</span> n_samples <span class="org-keyword">is</span> <span class="org-constant">None</span>:
      <span class="org-variable-name">n_samples</span> = torch.Size([1])
    <span class="org-keyword">if</span> torch.cuda.is_available():
      <span class="org-keyword">self</span>.cuda()
    <span class="org-variable-name">opt</span> = torch.optim.RMSprop(<span class="org-keyword">self</span>.parameters(), **kwargs)
    <span class="org-variable-name">global_step</span> = 0
    <span class="org-keyword">for</span> epoch <span class="org-keyword">in</span> <span class="org-builtin">range</span>(n_epochs):
      <span class="org-keyword">for</span> x, y <span class="org-keyword">in</span> data:
        <span class="org-keyword">if</span> torch.cuda.is_available():
          <span class="org-variable-name">x</span> = x.cuda()
        opt.zero_grad()
        <span class="org-variable-name">loss</span> = <span class="org-keyword">self</span>.forward(x, n_samples, writer, global_step)
        loss.backward()
        opt.step()
        <span class="org-variable-name">global_step</span> += 1
    <span class="org-keyword">return</span> <span class="org-keyword">self</span>
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">batch_size</span> = 32
<span class="org-variable-name">mnist_train</span> = td.DataLoader(
  tv.datasets.MNIST(
    root=<span class="org-string">'/scratch/midway2/aksarkar/singlecell/'</span>,
    transform=<span class="org-keyword">lambda</span> x: (np.frombuffer(x.tobytes(), dtype=<span class="org-string">'uint8'</span>) &gt; 0).astype(np.float32).reshape((1, x.size[0], x.size[1]))),
  batch_size=batch_size,
  shuffle=<span class="org-constant">True</span>,
  pin_memory=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">latent_dim</span> = 4
<span class="org-variable-name">n_epochs</span> = 2
<span class="org-variable-name">run</span> = 5
<span class="org-variable-name">model</span> = ConvVAE(input_h=28, input_w=28, latent_dim=latent_dim)
model.fit(mnist_train,
          n_epochs=n_epochs,
          lr=1e-5,
          log_dir=f<span class="org-string">'runs/convvae/mnist/latent_{latent_dim}_epochs_{n_epochs}_run_{run}'</span>)
</pre>
</div>

<pre class="example">
ConvVAE(
(encoder): Encoder(
(net): Sequential(
(0): Conv2d(1, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
(1): ReLU()
(2): Conv2d(32, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
(3): ReLU()
(4): Flatten()
(5): Linear(in_features=50176, out_features=1024, bias=True)
(6): ReLU()
(7): Linear(in_features=1024, out_features=512, bias=True)
(8): ReLU()
)
(mean): Linear(in_features=512, out_features=4, bias=True)
(scale): Sequential(
(0): Linear(in_features=512, out_features=4, bias=True)
(1): Softplus(beta=1, threshold=20)
)
)
(decoder): Decoder(
(net1): Sequential(
(0): Linear(in_features=4, out_features=512, bias=True)
(1): ReLU()
(2): Linear(in_features=512, out_features=1024, bias=True)
(3): ReLU()
(4): Linear(in_features=1024, out_features=50176, bias=True)
(5): ReLU()
)
(net2): Sequential(
(0): ConvTranspose2d(64, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
(1): ReLU()
(2): ConvTranspose2d(32, 1, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
(3): Sigmoid()
)
)
)
</pre>

<p>
Draw some samples.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">rng</span> = np.random.default_rng(1)
<span class="org-keyword">with</span> torch.no_grad():
  <span class="org-variable-name">z</span> = rng.normal(size=(10, latent_dim))
  <span class="org-variable-name">xh</span> = model.decoder.forward(torch.tensor(z, dtype=torch.<span class="org-builtin">float</span>, device=<span class="org-string">'cuda'</span>)).squeeze().cpu().numpy()
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 5)
<span class="org-keyword">for</span> i, a <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(ax.ravel()):
  a.imshow(xh[i], cmap=colorcet.cm[<span class="org-string">'gray'</span>])
  a.set_xticks([])
  a.set_yticks([])
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/spatial-vae.org/mnist-ex.png" alt="mnist-ex.png">
</p>
</div>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2020-10-01 Thu 17:04</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
