<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2018-09-12 Wed 16:50 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Multivariate fine mapping to identify target genes</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="https://www.pirilampo.org/styles/readtheorg/css/htmlize.css"/>
<style type="text/css">body {width: 60em; margin:auto} pre.src {overflow:auto}</style>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
<script type="text/x-mathjax-config">
    MathJax.Hub.Config({
        displayAlign: "center",
        displayIndent: "0em",

        "HTML-CSS": { scale: 100,
                        linebreaks: { automatic: "false" },
                        webFont: "TeX"
                       },
        SVG: {scale: 100,
              linebreaks: { automatic: "false" },
              font: "TeX"},
        NativeMML: {scale: 100},
        TeX: { equationNumbers: {autoNumber: "AMS"},
               MultLineWidth: "85%",
               TagSide: "right",
               TagIndent: ".8em"
             }
});
</script>
<script type="text/javascript"
        src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div id="content">
<h1 class="title">Multivariate fine mapping to identify target genes</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org5ab2471">Introduction</a></li>
<li><a href="#org1029da4">Methods</a>
<ul>
<li><a href="#org6a5da6e">Two step fine-mapping approach</a></li>
<li><a href="#org4b8ab0d">EP approach</a></li>
</ul>
</li>
<li><a href="#orgee4d4bf">Results</a>
<ul>
<li><a href="#orga1d0dba">Toy examples</a></li>
<li><a href="#org28c0c69">Simulation study</a></li>
<li><a href="#org8b5b2fb">Application to GTEx</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-org5ab2471" class="outline-2">
<h2 id="org5ab2471">Introduction</h2>
<div class="outline-text-2" id="text-org5ab2471">
<p>
One of the key challenges in interpreting non-coding variation is linking
regulatory variants to their target genes. One natural approach to solve this
problem is QTL mapping: directly associating genetic variation with gene
expression levels. QTL studies have revealed that genetic variants associated
with expression of a gene cluster near the transcription start site of that
gene (<a href="http://dx.doi.org/10.1371/journal.pgen.1000214">Veyrieras et al. 2008</a>, <a href="http://dx.doi.org/10.1126/science.1174148">Dimas et al. 2009</a>, <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1002639">Stranger et al. 2012</a>). Large
scale efforts such as the Gene Tissue Expression Project (GTEx) have
systematically profiled gene expression across cell types, and sought to use
the insights from QTL mapping to make progress on interrogating the genetic
basis of disease (<a href="https://www.nature.com/articles/nature24277">GTEx consortium 2017</a>).
</p>

<p>
However, recent efforts to experimentally fine map GWAS loci have shown that
causal non-coding variants can be more than 1 megabase from their target
gene, and may not even target the closest gene (<a href="https://www.nature.com/articles/nature09266">Musunru et al 2010</a>,
<a href="https://www.nejm.org/doi/full/10.1056/NEJMoa1502214">Claussnitzer et al. 2015</a>, <a href="https://elifesciences.org/articles/10557">Wang et al 2016</a>). Further, enhancer-gene
interactions have been shown to be limited by topologically associated domain
boundaries, rather than genomic distance (<a href="https://www.cell.com/cell/fulltext/S0092-8674(15)00377-3">Lupiáñes et al. 2015</a>). These
handful of examples are supported by systematic profiling of chromatin
interactions (<a href="https://www.nature.com/articles/nature12716">Zhang et al. 2013</a>) and high throughput perturbation assays
(<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5438575/">Fulco et al. 2016</a>).
</p>

<p>
Here, we develop a new method to fine map QTLs which does not assume that
SNPs target the nearest gene, or are within 1 MB of their target gene. The
key idea of our approach is to develop a novel multivariate fine mapping
approach under the prior assumption that any single SNP can have a non-zero
effect on at most one phenotype.
</p>
</div>
</div>

<div id="outline-container-org1029da4" class="outline-2">
<h2 id="org1029da4">Methods</h2>
<div class="outline-text-2" id="text-org1029da4">
</div>
<div id="outline-container-org6a5da6e" class="outline-3">
<h3 id="org6a5da6e">Two step fine-mapping approach</h3>
<div class="outline-text-3" id="text-org6a5da6e">
<p>
Suppose we assume a generative model with exactly one causal variant,
denoted <i>single effect regression</i> (SER):
</p>

<p>
\[ \mathbf{y} = \mathbf{X b} + \mathbf{e} \]
</p>

<p>
\[ b = \mathbf{z} \beta \]
</p>

<p>
\[ \mathbf{z} \sim \mathrm{Mult}(1, \mathbf{\pi}) \]
</p>

<p>
Then, posterior computations are easy:
</p>

<p>
\[ \mathbf{z} \mid \mathbf{X}, \mathbf{y} \sim \mathrm{Mult}(1, \mathbf{\alpha}) \]
</p>

<p>
\[ \alpha_j = \frac{\pi_j BF_j}{\sum_k \pi_k BF_k} \]
</p>

<p>
To allow more than one causal variant, a typical approach is to change the
prior, factorizing \(\gamma_1, \ldots, \gamma_p\). In <i>susie</i>, we take a
different approach:
</p>

<p>
\[ \mathbf{y} = \sum_l \mathbf{X} \mathbf{b}_l + \mathbf{e} \]
</p>

<p>
\[ b = \mathbf{z}_l \beta_l \]
</p>

<p>
In this approach, \(\mathbf{z_l}\) does not factorize, and the posterior
distribution reflects dependencies between the different variables. 
</p>

<p>
This model can be efficiently fit using variational inference, for which the
updates outline a Bayesian analogue of stepwise regression.
</p>

<p>
To generalize SER to <i>multivariate single effect regression</i> (MSER), we need
to assume a multivariate prior on \(\beta\):
</p>

<p>
\[ \mathbf{y}_k = \mathbf{X} \mathbf{b}_k + \mathbf{e} \]
</p>

<p>
\[ \mathbf{b}_k = \mathbf{z}_k \beta_k \]
</p>

<p>
\[ \mathbf{\beta} \sim g(\cdot) \]
</p>

<p>
If we assume \(\beta \sim \mathcal{N}(0, \tau^{-1} \mathbf{I})\), then we
get \(K\) independent SER problems. We can generalize MSER to multiviarate
<i>susie</i> by assuming multiple independent effects as above.
</p>

<p>
Now, we want to incorporate a further constraint, that each SNP can only be
causal for one phenotype. Suppose we fit multivariate <i>susie</i> on a genomic
region with \(K\) genes. For SNP \(j\), gene \(k\), we estimate
\(\lambda_{kj} = \mathrm{logit}(p(z_{kj} = 1 \mid \mathbf{X},
   \mathbf{Y}))\).
</p>

<p>
Fix SNP \(j\) and consider \(\lambda_{j} = (\lambda_{1j}, \ldots,
   \lambda_{Kj})\). If only one entry of \(\lambda_j\) is non-trivial, it
clearly satisfies the constraint. 
</p>

<p>
If more than one entry is, then we need to solve a sparse recovery problem:
we need to infer which of the signals is truly non-zero. Intuitively, having
solved this problem, we should go back and refine the solution to the
multivariate fine mapping.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">susie_mg</span>(X, Y, **kwargs):
  <span class="org-variable-name">pip</span> = []
  <span class="org-keyword">for</span> y <span class="org-keyword">in</span> Y.T:
    <span class="org-variable-name">res</span> = susier.susie(X, y.reshape(-1, 1), **kwargs)
    pip.append(np.array(susier.susie_get_PIP(res)))
  <span class="org-keyword">return</span> np.array(pip)
</pre>
</div>
</div>
</div>

<div id="outline-container-org4b8ab0d" class="outline-3">
<h3 id="org4b8ab0d">EP approach</h3>
<div class="outline-text-3" id="text-org4b8ab0d">
<p>
In SER, for each phenotype, we encoded the constraint that a single variant
is causal by assuming the prior \(\mathbf{z} \sim \mathrm{Mult}(1,
   \mathbf{\pi})\).
</p>

<p>
Ideally, in the multivariate case we would encode the constraint that each
SNP can only be causal for one phenotype in the same way. Suppose we only
had two phenotypes, only two SNPs, and only one effect per phenotype. We
need the following to hold simultaneously:
</p>

<ol class="org-ol">
<li>\(z_{11}, z_{12} \sim \mathrm{Mult}(1, \mathbf{\pi}_a)\)</li>
<li>\(z_{21}, z_{22} \sim \mathrm{Mult}(1, \mathbf{\pi}_a)\)</li>
<li>\(z_{11}, z_{21} \sim \mathrm{Mult}(1, \mathbf{\pi}_b)\)</li>
<li>\(z_{12}, z_{22} \sim \mathrm{Mult}(1, \mathbf{\pi}_b)\)</li>
</ol>

<p>
The resulting distribution on \(\mathbf{Z}\) is most readily expressed as a
factor graph:
</p>

<p>
\[ p(\mathbf{Z} \mid \cdot) \propto \prod_j f_j(z_{1j}, z_{2j}) \prod_k g_k(z_{k1}, z_{k2}) \]
</p>

<p>
The remainder of the model is easily converted into a factor graph:
</p>

<p>
\[ p(\cdot) \propto \prod_k p(Y_k \mid X_k, B_k, \tau) p(B \mid Z, \tau_0)
   p(Z \mid \pi_a, \pi_b) \]
</p>

<p>
We use expectation propagation to perform approximate inference on this
model (<a href="https://tminka.github.io/papers/ep/">Minka 2001</a>, <a href="10.1007/s10994-014-5475-7">Hernández-Lobato et al 2015</a>).
</p>
</div>
</div>
</div>

<div id="outline-container-orgee4d4bf" class="outline-2">
<h2 id="orgee4d4bf">Results</h2>
<div class="outline-text-2" id="text-orgee4d4bf">
</div>
<div id="outline-container-orga1d0dba" class="outline-3">
<h3 id="orga1d0dba">Toy examples</h3>
<div class="outline-text-3" id="text-orga1d0dba">
<p>
Consider just two genes, and two SNPs. For simplicity, generate bivariate
normal "genotypes":
</p>

<p>
\[ \left[\begin{array}{l}x_{i1}\\x_{i2}\end{array}\right] \sim
   \mathcal{N}\left(\mathbf{0}, \left[\begin{array}{ll} 1 &\rho\\ \rho &
   1\end{array}\right]\right) \]
</p>

<p>
\[ y_{i1} = x_{i1} \beta_1 + e_{i1} \]
</p>

<p>
\[ y_{i2} = x_{i2} \beta_2 + e_{i2} \]
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">simulate_ser_y</span>(x, pve):
  <span class="org-variable-name">n</span> = x.shape[0]
  <span class="org-variable-name">beta</span> = np.random.normal()
  <span class="org-variable-name">y</span> = x * beta
  <span class="org-variable-name">y</span> += np.random.normal(scale=np.sqrt(y.var() * (1 / pve - 1)), size=n)
  <span class="org-variable-name">y</span> -= y.mean(axis=0)
  <span class="org-keyword">return</span> y

<span class="org-keyword">def</span> <span class="org-function-name">simulate_ex1</span>(rho, n, pve=0.1, seed=0):
  np.random.seed(seed)
  <span class="org-variable-name">cov</span> = np.array([[1, rho], [rho, 1]])
  <span class="org-variable-name">x</span> = np.random.multivariate_normal(mean=np.zeros(2), cov=cov, size=n)
  <span class="org-variable-name">y1</span> = simulate_ser_y(x[:,0], pve)
  <span class="org-variable-name">y2</span> = simulate_ser_y(x[:,1], pve)
  <span class="org-variable-name">x</span> -= x.mean(axis=0)
  <span class="org-keyword">return</span> x, np.vstack([y1, y2]).T
</pre>
</div>

<p>
When genotypes are uncorrelated, we should expect <i>susie</i> to exactly recover
the causal SNP for each gene, and conversely the target gene for each SNP.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = simulate_ex1(rho=0, n=100)
susie_mg(X, Y, L=1)
</pre>
</div>

<pre class="example">
array([[0.97088521, 0.02911479],
[0.04866922, 0.95133078]])
</pre>

<p>
When genotypes are nearly perfectly correlated, the expression will become
nearly perfectly correlated. We should expect <i>susie</i> to put equal posterior
probability on each SNP, and therefore be unable to find the target gene.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = simulate_ex1(rho=0.99, n=100)
susie_mg(X, Y, L=1)
</pre>
</div>

<pre class="example">
array([[0.37021302, 0.62978698],
[0.6153817 , 0.3846183 ]])
</pre>

<p>
Surprisingly, <i>susie</i> simply gives the wrong answer in this case. Look at
the distribution of \(p(z_{11} \mid \cdot)\) over many trials.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">get_pip_11</span>(**kwargs):
  <span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = simulate_ex1(**kwargs)
  <span class="org-variable-name">res</span> = susie_mg(X, Y, L=1)
  <span class="org-keyword">return</span> res[0, 0]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">data</span> = [get_pip_11(rho=0.99, n=100, seed=i) <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(100)]

plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">f</span> = st.gaussian_kde(data)
<span class="org-variable-name">grid</span> = np.linspace(0, 1, 1000)
plt.plot(grid, f(grid), c=<span class="org-string">'k'</span>, lw=1)
plt.xlabel(<span class="org-string">'PIP of SNP 1 on gene 1'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'Density'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/susie.org/susie-mg-toy-pip.png" alt="susie-mg-toy-pip.png">
</p>
</div>

<p>
Fixing \(\rho\), as the number of samples \(n\) increases, <i>susie</i> gets the
correct answer on average, but there is still considerable variation.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">pip_vs_n</span> = pd.DataFrame([(<span class="org-builtin">int</span>(n), i, get_pip_11(rho=0.99, n=<span class="org-builtin">int</span>(n), seed=i))
                         <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(50) <span class="org-keyword">for</span> n <span class="org-keyword">in</span> np.logspace(2, 4, 100)],
                        columns=[<span class="org-string">'n'</span>, <span class="org-string">'trial'</span>, <span class="org-string">'pip_11'</span>])
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
plt.semilogx()
plt.ylim(0, 1)
plt.axhline(y=0.5, c=<span class="org-string">'0.5'</span>, lw=1, ls=<span class="org-string">'--'</span>)
<span class="org-variable-name">m</span> = pip_vs_n.groupby(<span class="org-string">'n'</span>)[<span class="org-string">'pip_11'</span>].agg(np.mean)
<span class="org-variable-name">s</span> = pip_vs_n.groupby(<span class="org-string">'n'</span>)[<span class="org-string">'pip_11'</span>].agg(np.std)
plt.plot(np.logspace(2, 4, 100), m, lw=1, c=<span class="org-string">'k'</span>)
plt.fill_between(np.logspace(2, 4, 100), m - 1.96 * s, m + 1.96 * s, color=<span class="org-string">'k'</span>, alpha=0.1)
plt.xlabel(<span class="org-string">'Sample size'</span>)
<span class="org-variable-name">_</span> = plt.ylabel(<span class="org-string">'PIP of SNP 1 on gene 1'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/susie.org/susie-mg-toy-vs-n.png" alt="susie-mg-toy-vs-n.png">
</p>
</div>

<p>
Now, consider the case of uncorrelated genotypes, but correlated
expression. The only way this can arise is if some unobserved factor
influences both genes.
</p>

<p>
\[ \mathbf{X}_i \sim \mathcal{N}(\mathbf{0}, \mathbf{I}) \]
</p>

<p>
\[ u_i \sim \mathcal{N}(0, \sigma^2_u) \]
</p>

<p>
\[ \beta_k \sim \mathcal{N}(0, \sigma^2_g) \]
</p>

<p>
\[ \mathbf{Y} = \mathbf{X}\beta + \left[\mathbf{u}\ \mathbf{u}\right] +
   \mathbf{E} \]
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">simulate_ser_u_y</span>(x, u, pve_x, pve_u):
  <span class="org-variable-name">n</span> = x.shape[0]
  <span class="org-variable-name">beta</span> = np.random.normal(scale=np.sqrt(pve_x / x.var()))
  <span class="org-variable-name">y</span> = x * beta + u
  <span class="org-variable-name">y</span> += np.random.normal(scale=np.sqrt(y.var() * (1 / (pve_x + pve_u) - 1)), size=n)
  <span class="org-variable-name">y</span> -= y.mean(axis=0)
  <span class="org-keyword">return</span> y

<span class="org-keyword">def</span> <span class="org-function-name">simulate_ex2</span>(n, pve_x=0.1, pve_u=0.1, seed=0):
  np.random.seed(seed)
  <span class="org-variable-name">x</span> = np.random.multivariate_normal(mean=np.zeros(2), cov=np.eye(2), size=n)
  <span class="org-variable-name">u</span> = np.random.normal(scale=np.sqrt(pve_u), size=n)
  <span class="org-variable-name">y1</span> = simulate_ser_u_y(x[:,0], u, pve_x, pve_u)
  <span class="org-variable-name">y2</span> = simulate_ser_u_y(x[:,1], u, pve_x, pve_u)
  <span class="org-variable-name">x</span> -= x.mean(axis=0)
  <span class="org-keyword">return</span> x, np.vstack([y1, y2]).T
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">X</span>, <span class="org-variable-name">Y</span> = simulate_ex2(n=100, pve_x=.05, pve_u=.9)
susie_mg(X, Y, L=1)
</pre>
</div>

<pre class="example">
array([[0.92064932, 0.07935068],
[0.0102968 , 0.9897032 ]])
</pre>

<div class="org-src-container">
<pre class="src src-ipython">np.corrcoef(Y.T)
</pre>
</div>

<pre class="example">
array([[1.       , 0.8488416],
[0.8488416, 1.       ]])
</pre>

<p>
Even when expression is highly correlated due to some unobserved factor
(e.g., <i>trans</i>-QTL), <i>susie</i> can still recover the <i>cis</i>-eQTLs.
</p>

<p>
Overall, the results suggest that simply applying univariate <i>susie</i> over
large windows will get the correct answer whenever it is possible to get the
correct answer.
</p>
</div>
</div>

<div id="outline-container-org28c0c69" class="outline-3">
<h3 id="org28c0c69">Simulation study</h3>
</div>
<div id="outline-container-org8b5b2fb" class="outline-3">
<h3 id="org8b5b2fb">Application to GTEx</h3>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2018-09-12 Wed 16:50</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
