<!DOCTYPE html>
<html lang="en">
<head>
<!-- 2021-08-25 Wed 18:52 -->
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1">
<title>Cardiac trajectory prediction</title>
<meta name="generator" content="Org mode">
<meta name="author" content="Abhishek Sarkar">
<style type="text/css">
 <!--/*--><![CDATA[/*><!--*/
  .title  { text-align: center;
             margin-bottom: .2em; }
  .subtitle { text-align: center;
              font-size: medium;
              font-weight: bold;
              margin-top:0; }
  .todo   { font-family: monospace; color: red; }
  .done   { font-family: monospace; color: green; }
  .priority { font-family: monospace; color: orange; }
  .tag    { background-color: #eee; font-family: monospace;
            padding: 2px; font-size: 80%; font-weight: normal; }
  .timestamp { color: #bebebe; }
  .timestamp-kwd { color: #5f9ea0; }
  .org-right  { margin-left: auto; margin-right: 0px;  text-align: right; }
  .org-left   { margin-left: 0px;  margin-right: auto; text-align: left; }
  .org-center { margin-left: auto; margin-right: auto; text-align: center; }
  .underline { text-decoration: underline; }
  #postamble p, #preamble p { font-size: 90%; margin: .2em; }
  p.verse { margin-left: 3%; }
  pre {
    border: 1px solid #ccc;
    box-shadow: 3px 3px 3px #eee;
    padding: 8pt;
    font-family: monospace;
    overflow: auto;
    margin: 1.2em;
  }
  pre.src {
    position: relative;
    overflow: visible;
    padding-top: 1.2em;
  }
  pre.src:before {
    display: none;
    position: absolute;
    background-color: white;
    top: -10px;
    right: 10px;
    padding: 3px;
    border: 1px solid black;
  }
  pre.src:hover:before { display: inline;}
  /* Languages per Org manual */
  pre.src-asymptote:before { content: 'Asymptote'; }
  pre.src-awk:before { content: 'Awk'; }
  pre.src-C:before { content: 'C'; }
  /* pre.src-C++ doesn't work in CSS */
  pre.src-clojure:before { content: 'Clojure'; }
  pre.src-css:before { content: 'CSS'; }
  pre.src-D:before { content: 'D'; }
  pre.src-ditaa:before { content: 'ditaa'; }
  pre.src-dot:before { content: 'Graphviz'; }
  pre.src-calc:before { content: 'Emacs Calc'; }
  pre.src-emacs-lisp:before { content: 'Emacs Lisp'; }
  pre.src-fortran:before { content: 'Fortran'; }
  pre.src-gnuplot:before { content: 'gnuplot'; }
  pre.src-haskell:before { content: 'Haskell'; }
  pre.src-hledger:before { content: 'hledger'; }
  pre.src-java:before { content: 'Java'; }
  pre.src-js:before { content: 'Javascript'; }
  pre.src-latex:before { content: 'LaTeX'; }
  pre.src-ledger:before { content: 'Ledger'; }
  pre.src-lisp:before { content: 'Lisp'; }
  pre.src-lilypond:before { content: 'Lilypond'; }
  pre.src-lua:before { content: 'Lua'; }
  pre.src-matlab:before { content: 'MATLAB'; }
  pre.src-mscgen:before { content: 'Mscgen'; }
  pre.src-ocaml:before { content: 'Objective Caml'; }
  pre.src-octave:before { content: 'Octave'; }
  pre.src-org:before { content: 'Org mode'; }
  pre.src-oz:before { content: 'OZ'; }
  pre.src-plantuml:before { content: 'Plantuml'; }
  pre.src-processing:before { content: 'Processing.js'; }
  pre.src-python:before { content: 'Python'; }
  pre.src-R:before { content: 'R'; }
  pre.src-ruby:before { content: 'Ruby'; }
  pre.src-sass:before { content: 'Sass'; }
  pre.src-scheme:before { content: 'Scheme'; }
  pre.src-screen:before { content: 'Gnu Screen'; }
  pre.src-sed:before { content: 'Sed'; }
  pre.src-sh:before { content: 'shell'; }
  pre.src-sql:before { content: 'SQL'; }
  pre.src-sqlite:before { content: 'SQLite'; }
  /* additional languages in org.el's org-babel-load-languages alist */
  pre.src-forth:before { content: 'Forth'; }
  pre.src-io:before { content: 'IO'; }
  pre.src-J:before { content: 'J'; }
  pre.src-makefile:before { content: 'Makefile'; }
  pre.src-maxima:before { content: 'Maxima'; }
  pre.src-perl:before { content: 'Perl'; }
  pre.src-picolisp:before { content: 'Pico Lisp'; }
  pre.src-scala:before { content: 'Scala'; }
  pre.src-shell:before { content: 'Shell Script'; }
  pre.src-ebnf2ps:before { content: 'ebfn2ps'; }
  /* additional language identifiers per "defun org-babel-execute"
       in ob-*.el */
  pre.src-cpp:before  { content: 'C++'; }
  pre.src-abc:before  { content: 'ABC'; }
  pre.src-coq:before  { content: 'Coq'; }
  pre.src-groovy:before  { content: 'Groovy'; }
  /* additional language identifiers from org-babel-shell-names in
     ob-shell.el: ob-shell is the only babel language using a lambda to put
     the execution function name together. */
  pre.src-bash:before  { content: 'bash'; }
  pre.src-csh:before  { content: 'csh'; }
  pre.src-ash:before  { content: 'ash'; }
  pre.src-dash:before  { content: 'dash'; }
  pre.src-ksh:before  { content: 'ksh'; }
  pre.src-mksh:before  { content: 'mksh'; }
  pre.src-posh:before  { content: 'posh'; }
  /* Additional Emacs modes also supported by the LaTeX listings package */
  pre.src-ada:before { content: 'Ada'; }
  pre.src-asm:before { content: 'Assembler'; }
  pre.src-caml:before { content: 'Caml'; }
  pre.src-delphi:before { content: 'Delphi'; }
  pre.src-html:before { content: 'HTML'; }
  pre.src-idl:before { content: 'IDL'; }
  pre.src-mercury:before { content: 'Mercury'; }
  pre.src-metapost:before { content: 'MetaPost'; }
  pre.src-modula-2:before { content: 'Modula-2'; }
  pre.src-pascal:before { content: 'Pascal'; }
  pre.src-ps:before { content: 'PostScript'; }
  pre.src-prolog:before { content: 'Prolog'; }
  pre.src-simula:before { content: 'Simula'; }
  pre.src-tcl:before { content: 'tcl'; }
  pre.src-tex:before { content: 'TeX'; }
  pre.src-plain-tex:before { content: 'Plain TeX'; }
  pre.src-verilog:before { content: 'Verilog'; }
  pre.src-vhdl:before { content: 'VHDL'; }
  pre.src-xml:before { content: 'XML'; }
  pre.src-nxml:before { content: 'XML'; }
  /* add a generic configuration mode; LaTeX export needs an additional
     (add-to-list 'org-latex-listings-langs '(conf " ")) in .emacs */
  pre.src-conf:before { content: 'Configuration File'; }

  table { border-collapse:collapse; }
  caption.t-above { caption-side: top; }
  caption.t-bottom { caption-side: bottom; }
  td, th { vertical-align:top;  }
  th.org-right  { text-align: center;  }
  th.org-left   { text-align: center;   }
  th.org-center { text-align: center; }
  td.org-right  { text-align: right;  }
  td.org-left   { text-align: left;   }
  td.org-center { text-align: center; }
  dt { font-weight: bold; }
  .footpara { display: inline; }
  .footdef  { margin-bottom: 1em; }
  .figure { padding: 1em; }
  .figure p { text-align: center; }
  .inlinetask {
    padding: 10px;
    border: 2px solid gray;
    margin: 10px;
    background: #ffffcc;
  }
  #org-div-home-and-up
   { text-align: right; font-size: 70%; white-space: nowrap; }
  textarea { overflow-x: auto; }
  .linenr { font-size: smaller }
  .code-highlighted { background-color: #ffff00; }
  .org-info-js_info-navigation { border-style: none; }
  #org-info-js_console-label
    { font-size: 10px; font-weight: bold; white-space: nowrap; }
  .org-info-js_search-highlight
    { background-color: #ffff00; color: #000000; font-weight: bold; }
  .org-svg { width: 90%; }
  /*]]>*/-->
</style>
<link href="bootstrap.min.css" rel="stylesheet" />
<link rel="stylesheet" type="text/css" href="htmlize.css"/>
<link rel="stylesheet" type="text/css" href="main.css"/>
<script type="text/javascript">
/*
@licstart  The following is the entire license notice for the
JavaScript code in this tag.

Copyright (C) 2012-2017 Free Software Foundation, Inc.

The JavaScript code in this tag is free software: you can
redistribute it and/or modify it under the terms of the GNU
General Public License (GNU GPL) as published by the Free Software
Foundation, either version 3 of the License, or (at your option)
any later version.  The code is distributed WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.  See the GNU GPL for more details.

As additional permission under GNU GPL version 3 section 7, you
may distribute non-source (e.g., minimized or compacted) forms of
that code without the copy of the GNU GPL normally required by
section 4, provided you include this license notice and a URL
through which recipients can access the Corresponding Source.


@licend  The above is the entire license notice
for the JavaScript code in this tag.
*/
<!--/*--><![CDATA[/*><!--*/
 function CodeHighlightOn(elem, id)
 {
   var target = document.getElementById(id);
   if(null != target) {
     elem.cacheClassElem = elem.className;
     elem.cacheClassTarget = target.className;
     target.className = "code-highlighted";
     elem.className   = "code-highlighted";
   }
 }
 function CodeHighlightOff(elem, id)
 {
   var target = document.getElementById(id);
   if(elem.cacheClassElem)
     elem.className = elem.cacheClassElem;
   if(elem.cacheClassTarget)
     target.className = elem.cacheClassTarget;
 }
/*]]>*///-->
</script>
</head>
<body>
<div id="content">
<h1 class="title">Cardiac trajectory prediction</h1>
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#orgf17561b">Introduction</a></li>
<li><a href="#orgd5fe55c">Setup</a></li>
<li><a href="#orgcfb7f09">Materials and methods</a>
<ul>
<li><a href="#orgca3dd86">Selewa et al. iPSC-CM data</a></li>
<li><a href="#org41e4dbb">Elorbany et al. iPSC-CM data</a></li>
</ul>
</li>
<li><a href="#org2a6603e">Results</a>
<ul>
<li><a href="#orga00ec74">Diffusion pseudotime</a></li>
<li><a href="#orgbddb008">Dependence of DPT on preprocessing</a></li>
<li><a href="#org1c0b384">NMF-based pseudotime</a></li>
<li><a href="#orgdef75dc">NMF pseudotime</a></li>
<li><a href="#org80fe738">Per individual analysis</a></li>
<li><a href="#orgf6fa745">Geodesics</a></li>
</ul>
</li>
</ul>
</div>
</div>

<div id="outline-container-orgf17561b" class="outline-2">
<h2 id="orgf17561b">Introduction</h2>
<div class="outline-text-2" id="text-orgf17561b">
<p>
Here, we investigate some ideas for modeling differentiation of iPSCs into
cardiomyocytes. Our goal is to build intuition about methods which model gene
expression trajectories.
</p>

<p>
We revisit this in light of a recent publication
(<a href="https://www.nature.com/articles/s41588-021-00894-z">Morabito et
al. 2021</a>), that used a recurrent VAE
(<a href="https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/btab260/6273572">Mitra
and MacLean 2021</a>). The input to the VAE was diffusion pseudotime (DPT),
which seems suspect given our preliminary results on the uncertainty of DPT.
</p>
</div>
</div>

<div id="outline-container-orgd5fe55c" class="outline-2">
<h2 id="orgd5fe55c">Setup</h2>
<div class="outline-text-2" id="text-orgd5fe55c">
<div class="org-src-container">
<pre class="src src-ipython" id="org9f8bc1c"><span class="org-keyword">import</span> anndata
<span class="org-keyword">import</span> collections
<span class="org-keyword">import</span> functools <span class="org-keyword">as</span> ft
<span class="org-keyword">import</span> hyperopt <span class="org-keyword">as</span> hp
<span class="org-keyword">import</span> mpebpm
<span class="org-keyword">import</span> numpy <span class="org-keyword">as</span> np
<span class="org-keyword">import</span> pandas <span class="org-keyword">as</span> pd
<span class="org-keyword">import</span> scanpy <span class="org-keyword">as</span> sc
<span class="org-keyword">import</span> scipy.sparse <span class="org-keyword">as</span> ss
<span class="org-keyword">import</span> scipy.special <span class="org-keyword">as</span> sp
<span class="org-keyword">import</span> scipy.stats <span class="org-keyword">as</span> st
<span class="org-keyword">import</span> torch
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> rpy2.robjects.packages
<span class="org-keyword">import</span> rpy2.robjects.pandas2ri
rpy2.robjects.pandas2ri.activate()
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">%matplotlib inline
%config <span class="org-variable-name">InlineBackend.figure_formats</span> = <span class="org-builtin">set</span>([<span class="org-string">'retina'</span>])
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> colorcet
<span class="org-keyword">import</span> matplotlib
<span class="org-keyword">import</span> matplotlib.pyplot <span class="org-keyword">as</span> plt
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'figure.facecolor'</span>] = <span class="org-string">'w'</span>
<span class="org-variable-name">plt.rcParams</span>[<span class="org-string">'font.family'</span>] = <span class="org-string">'Nimbus Sans'</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgcfb7f09" class="outline-2">
<h2 id="orgcfb7f09">Materials and methods</h2>
<div class="outline-text-2" id="text-orgcfb7f09">
</div>
<div id="outline-container-orgca3dd86" class="outline-3">
<h3 id="orgca3dd86">Selewa et al. iPSC-CM data</h3>
<div class="outline-text-3" id="text-orgca3dd86">
<p>
<a href="https://www.nature.com/articles/s41598-020-58327-6">Selewa et al. 2020</a>
differentiated iPSCs derived from from two donor individuals into
cardiomyocytes, and performed Drop-Seq and DroNC-Seq at 4 and 5 time points,
respectively: 0, 1, 3, 7, and 15 days.
</p>

<p>
Read the data.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">raw</span> = pd.read_table(<span class="org-string">'/project2/onibasu/data/public_html/Alan_CZI/DGE_Matrices/DROP_combined_counts.tsv.gz'</span>, index_col=0)
</pre>
</div>

<p>
Convert to <code>adata</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = anndata.AnnData(ss.csc_matrix(raw.values.T),
                      obs=pd.DataFrame([pd.Series(x) <span class="org-keyword">for</span> x <span class="org-keyword">in</span> raw.columns.<span class="org-builtin">str</span>.split(<span class="org-string">'_'</span>)]),
                      var=raw.index.to_series().to_frame())
</pre>
</div>

<p>
Fix up the column names for <code>h5ad</code> output.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat.var.columns</span> = [<span class="org-string">'gene'</span>]
<span class="org-variable-name">dat.obs.columns</span> = [<span class="org-string">'barcode'</span>, <span class="org-string">'day'</span>, <span class="org-string">'donor'</span>, <span class="org-string">'dummy1'</span>, <span class="org-string">'dummy2'</span>]
</pre>
</div>

<p>
Drop genes with no non-zero observations.
</p>

<div class="org-src-container">
<pre class="src src-ipython">sc.pp.filter_genes(dat, min_cells=1)
dat.shape
</pre>
</div>

<pre class="example">
(20816, 24486)

</pre>

<p>
Write out to <code>h5ad</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython">dat.write(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/GSE129096_DROP_combined_allcells.h5ad'</span>)
</pre>
</div>

<p>
Read the <code>h5ad</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = anndata.read_h5ad(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/GSE129096_DROP_combined_allcells.h5ad'</span>)
</pre>
</div>
</div>
</div>

<div id="outline-container-org41e4dbb" class="outline-3">
<h3 id="org41e4dbb">Elorbany et al. iPSC-CM data</h3>
<div class="outline-text-3" id="text-org41e4dbb">
<p>
<a href="https://www.biorxiv.org/content/10.1101/2021.06.03.446970v1">Elorbany et
al. 2021</a> generated iPSC-CM differentiation scRNA-seq time course data in
19 cell lines at 7 time points. Read the data.
</p>

<div class="org-src-container">
<pre class="src src-R"><span class="org-ess-modifiers">library</span>(Seurat)
dat <span class="org-ess-assignment">&lt;-</span> readRDS(<span class="org-string">"/project2/gilad/reem/singlecellCM/Sobjs/xfinalmerged/Sobj_allrounds_finalmerged_noNA.rds"</span>)
saveRDS(dat$RNA@data, <span class="org-string">"/scratch/midway2/aksarkar/singlecell/elorbany-counts.Rds"</span>)
saveRDS(dat$RNA@meta.features, <span class="org-string">"/scratch/midway2/aksarkar/singlecell/elorbany-genes.Rds"</span>)
saveRDS(dat@meta.data, <span class="org-string">"/scratch/midway2/aksarkar/singlecell/elorbany-cells.Rds"</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">readRDS</span> = rpy2.robjects.r[<span class="org-string">'readRDS'</span>]
<span class="org-variable-name">x</span> = readRDS(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/elorbany-counts.Rds'</span>)
<span class="org-variable-name">cells</span> = readRDS(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/elorbany-cells.Rds'</span>)
rpy2.robjects.pandas2ri.deactivate()
<span class="org-variable-name">genes</span> = pd.Series(<span class="org-builtin">list</span>(readRDS(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/elorbany-genes.Rds'</span>).rownames)).to_frame()
rpy2.robjects.pandas2ri.activate()
</pre>
</div>

<p>
Convert to <code>adata</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat</span> = anndata.AnnData(
  ss.csc_matrix((x.slots[<span class="org-string">'x'</span>], x.slots[<span class="org-string">'i'</span>], x.slots[<span class="org-string">'p'</span>]), shape=x.slots[<span class="org-string">'Dim'</span>]).T,
  obs=cells,
  var=genes)
</pre>
</div>

<p>
Write to <code>h5ad</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">dat.var.columns</span> = [<span class="org-string">'name'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">dat.write_h5ad(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/elorbany-ipsc-cm.h5ad'</span>)
</pre>
</div>

<p>
Read the <code>h5ad</code>.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">elorbany_dat</span> = anndata.read_h5ad(<span class="org-string">'/scratch/midway2/aksarkar/singlecell/elorbany-ipsc-cm.h5ad'</span>)
</pre>
</div>
</div>
</div>
</div>

<div id="outline-container-org2a6603e" class="outline-2">
<h2 id="org2a6603e">Results</h2>
<div class="outline-text-2" id="text-org2a6603e">
</div>
<div id="outline-container-orga00ec74" class="outline-3">
<h3 id="orga00ec74">Diffusion pseudotime</h3>
<div class="outline-text-3" id="text-orga00ec74">
<p>
Compute DPT in the Selewa et al. data. By default, the method embeds the
data in principal component space to compute distances.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">reps</span> = {k: dat[dat.obs[<span class="org-string">'donor'</span>] == k] <span class="org-keyword">for</span> k <span class="org-keyword">in</span> (<span class="org-string">'Rep1'</span>, <span class="org-string">'Rep2'</span>)}
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> reps:
  sc.pp.filter_genes(reps[k], min_cells=1)
  sc.pp.pca(reps[k])
  sc.pp.neighbors(reps[k])
  sc.tl.diffmap(reps[k])
</pre>
</div>

<p>
Compute DPT in each donor, starting from 200 random starting points (day 0
cells).
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(1)
<span class="org-variable-name">dpt</span> = collections.defaultdict(<span class="org-builtin">list</span>)
<span class="org-keyword">for</span> k <span class="org-keyword">in</span> reps:
  <span class="org-keyword">for</span> root <span class="org-keyword">in</span> np.random.choice(reps[k].obs[reps[k].obs[<span class="org-string">'day'</span>] == <span class="org-string">'0'</span>].index, 200):
    reps[k]<span class="org-variable-name">.uns</span>[<span class="org-string">'iroot'</span>] = <span class="org-builtin">int</span>(root)
    sc.tl.dpt(reps[k], n_dcs=15)
    dpt[k].append(reps[k].obs[<span class="org-string">'dpt_pseudotime'</span>])
    <span class="org-keyword">del</span> reps[k].obs[<span class="org-string">'dpt_pseudotime'</span>]
</pre>
</div>

<p>
Order the cells in each donor by average DPT, then downsample and plot error
bars.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = plt.get_cmap(<span class="org-string">'Dark2'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(2, 1, sharex=<span class="org-constant">True</span>)
fig.set_size_inches(6, 4)
<span class="org-keyword">for</span> a, k <span class="org-keyword">in</span> <span class="org-builtin">zip</span>(ax, dpt):
  <span class="org-variable-name">temp</span> = np.array(dpt[k]).T
  <span class="org-variable-name">mean_dpt</span> = temp.mean(axis=1)
  <span class="org-variable-name">order</span> = np.argsort(mean_dpt)[::100]
  <span class="org-variable-name">n</span> = 0
  <span class="org-keyword">for</span> i, (day, _) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(reps[k].obs.groupby(<span class="org-string">'day'</span>)):
    <span class="org-variable-name">idx</span> = reps[k].obs[<span class="org-string">'day'</span>].iloc[order] == day
    a.errorbar(
      x=np.arange(order.shape[0])[idx],
      y=mean_dpt[order][idx],
      yerr=temp[order][idx].std(axis=1),
      fmt=<span class="org-string">'o'</span>,
      markersize=1,
      elinewidth=0.5,
      c=cm(i),
      label=day)
    <span class="org-variable-name">n</span> += <span class="org-builtin">len</span>(idx)
<span class="org-keyword">for</span> i, a <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(ax):
  a.set_title(f<span class="org-string">'Donor {i + 1}'</span>)
  a.set_ylim(0, 1)
  a.set_ylabel(<span class="org-string">'Diffusion pseudotime'</span>)
  <span class="org-variable-name">leg</span> = a.legend(frameon=<span class="org-constant">False</span>, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
  leg.set_title(<span class="org-string">'Day'</span>)
ax[-1].set_xlabel(<span class="org-string">'Cell (subsampled)'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/trajectory.org/drop-seq-dpt.png" alt="drop-seq-dpt.png">
</p>
</div>

<p>
Repeat the analysis for the Elorbany et al. data. Get the individual with
the most cells.
</p>

<div class="org-src-container">
<pre class="src src-ipython">elorbany_dat.obs.groupby([<span class="org-string">'individual'</span>])[<span class="org-string">'diffday'</span>].agg(<span class="org-builtin">len</span>).sort_values().tail()
</pre>
</div>

<pre class="example">
individual
NA18511    22203
NA18858    23197
NA19093    24299
NA19190    24405
NA18499    29466
Name: diffday, dtype: int64
</pre>

<p>
Compute DPT in that donor, starting from 200 random starting points (day 0
cells).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">temp</span> = elorbany_dat[elorbany_dat.obs[<span class="org-string">'individual'</span>] == <span class="org-string">'NA18499'</span>]
sc.pp.filter_genes(temp, min_cells=1)
sc.pp.pca(temp)
sc.pp.neighbors(temp, n_neighbors=30)
sc.tl.diffmap(temp, n_comps=10)

np.random.seed(1)
<span class="org-variable-name">dpt</span> = []
<span class="org-keyword">for</span> root <span class="org-keyword">in</span> np.random.choice(np.flatnonzero(temp.obs[<span class="org-string">'diffday'</span>] == <span class="org-string">'Day 0'</span>), 200):
  <span class="org-variable-name">temp.uns</span>[<span class="org-string">'iroot'</span>] = root
  sc.tl.dpt(temp, n_dcs=10)
  dpt.append(temp.obs[<span class="org-string">'dpt_pseudotime'</span>].values)
  <span class="org-keyword">del</span> temp.obs[<span class="org-string">'dpt_pseudotime'</span>]
<span class="org-variable-name">dpt</span> = np.array(dpt).T
</pre>
</div>

<p>
Order the cells by average DPT, then downsample and plot error bars.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">cm</span> = colorcet.cm[<span class="org-string">'bmy'</span>]
plt.clf()
plt.gcf().set_size_inches(6, 2.5)
<span class="org-variable-name">mean_dpt</span> = dpt.mean(axis=1)
<span class="org-variable-name">order</span> = np.argsort(mean_dpt)[::200]
<span class="org-keyword">for</span> i, (day, _) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(temp.obs.groupby(<span class="org-string">'diffday'</span>)):
  <span class="org-variable-name">idx</span> = temp.obs[<span class="org-string">'diffday'</span>].iloc[order] == day
  plt.errorbar(
    x=np.arange(order.shape[0])[idx],
    y=mean_dpt[order][idx],
    yerr=dpt[order][idx].std(axis=1),
    fmt=<span class="org-string">'o'</span>,
    markersize=1,
    elinewidth=0.5,
    c=cm(<span class="org-builtin">int</span>(day[4:]) / 15),
    label=day)
<span class="org-variable-name">leg</span> = plt.legend(frameon=<span class="org-constant">False</span>, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
leg.set_title(<span class="org-string">'Diffday'</span>)
plt.title(<span class="org-string">'NA18499'</span>)
plt.ylim(0, 1)
plt.ylabel(<span class="org-string">'Diffusion pseudotime'</span>)
plt.xlabel(<span class="org-string">'Cell (subsampled)'</span>)
plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/trajectory.org/elorbany-dpt.png" alt="elorbany-dpt.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgbddb008" class="outline-3">
<h3 id="orgbddb008">Dependence of DPT on preprocessing</h3>
<div class="outline-text-3" id="text-orgbddb008">
<p>
Selewa et al. write &ldquo;Selection of genes for trajectory analysis, or feature
selection, is critical for obtaining accurate trajectories.&rdquo; Since we know
something about the true ordering of cells, one can use model-based
optimization to find optimal pre-processing filters.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">def</span> <span class="org-function-name">loss</span>(theta, dat, n_samples=200):
  <span class="org-variable-name">min_cells</span>, <span class="org-variable-name">n_pcs</span>, <span class="org-variable-name">n_neighbors</span>, <span class="org-variable-name">n_dcs</span> = theta
  <span class="org-variable-name">temp</span> = dat.copy()
  sc.pp.filter_genes(temp, min_cells=<span class="org-builtin">int</span>(min_cells))
  <span class="org-comment-delimiter"># </span><span class="org-comment">Important: PCA is randomized too</span>
  np.random.seed(1)
  sc.pp.pca(temp, n_comps=<span class="org-builtin">int</span>(n_pcs))
  sc.pp.neighbors(temp, n_neighbors=<span class="org-builtin">int</span>(n_neighbors))
  sc.tl.diffmap(temp, n_comps=<span class="org-builtin">int</span>(n_dcs))
  <span class="org-variable-name">dpt</span> = []
  <span class="org-keyword">for</span> root <span class="org-keyword">in</span> np.random.choice(np.flatnonzero(temp.obs[<span class="org-string">'day'</span>] == <span class="org-string">'0'</span>), n_samples):
    <span class="org-variable-name">temp.uns</span>[<span class="org-string">'iroot'</span>] = root
    sc.tl.dpt(temp, n_dcs=<span class="org-builtin">int</span>(n_dcs))
    dpt.append(temp.obs[<span class="org-string">'dpt_pseudotime'</span>])
    <span class="org-keyword">del</span> temp.obs[<span class="org-string">'dpt_pseudotime'</span>]

  <span class="org-variable-name">expected_order</span> = pd.Categorical(dat.obs[<span class="org-string">'day'</span>].<span class="org-builtin">apply</span>(<span class="org-builtin">float</span>), ordered=<span class="org-constant">True</span>).codes
  <span class="org-variable-name">order</span> = np.argsort(np.array(dpt).T.mean(axis=1))
  <span class="org-variable-name">ranks</span> = np.empty_like(order)
  <span class="org-variable-name">ranks</span>[order] = np.arange(<span class="org-builtin">len</span>(order))
  <span class="org-keyword">return</span> {<span class="org-string">'loss'</span>: -st.kendalltau(ranks, expected_order)[0], <span class="org-string">'status'</span>: hp.STATUS_OK}
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">rep2</span> = dat[dat.obs[<span class="org-string">'donor'</span>] == <span class="org-string">'Rep2'</span>]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(1)
<span class="org-variable-name">trials</span> = hp.Trials()
<span class="org-variable-name">opt</span> = hp.fmin(
  ft.partial(loss, dat=rep2, n_samples=200),
  space=[hp.hp.loguniform(<span class="org-string">'min_cells'</span>, 0, np.log(1000)),
         hp.hp.uniform(<span class="org-string">'n_pcs'</span>, 3, 50),
         hp.hp.uniform(<span class="org-string">'n_neighbors'</span>, 5, 50),
         hp.hp.uniform(<span class="org-string">'n_dcs'</span>, 2, 50)],
  algo=hp.tpe.suggest,
  trials=trials,
  max_evals=100)
</pre>
</div>

<p>
0 - b238f6e9-37dd-4970-abeb-0eeb347a11ee
</p>
</div>
</div>

<div id="outline-container-org1c0b384" class="outline-3">
<h3 id="org1c0b384">NMF-based pseudotime</h3>
<div class="outline-text-3" id="text-org1c0b384">
<p>
Look at the diffusion maps.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">adata.obsm</span>[<span class="org-string">'X_nmf'</span>] = L
<span class="org-variable-name">rep1</span> = adata[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep1'</span>]
sc.pp.neighbors(rep1, use_rep=<span class="org-string">'X_nmf'</span>)
sc.tl.diffmap(rep1)

<span class="org-variable-name">rep2</span> = adata[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep2'</span>]
sc.pp.neighbors(rep2, use_rep=<span class="org-string">'X_nmf'</span>)
sc.tl.diffmap(rep2)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">from</span> mpl_toolkits.mplot3d <span class="org-keyword">import</span> Axes3D

plt.clf()
plt.gcf().set_size_inches(8, 5)

<span class="org-variable-name">ax</span> = plt.gcf().add_subplot(121, projection=<span class="org-string">'3d'</span>)
ax.grid(<span class="org-constant">False</span>)
<span class="org-keyword">for</span> i, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(rep1.obs.groupby(<span class="org-string">'day'</span>)):
  ax.scatter(rep1.obsm[<span class="org-string">'X_diffmap'</span>][rep1.obs[<span class="org-string">'day'</span>] == k,1],
              rep1.obsm[<span class="org-string">'X_diffmap'</span>][rep1.obs[<span class="org-string">'day'</span>] == k,2],
              rep1.obsm[<span class="org-string">'X_diffmap'</span>][rep1.obs[<span class="org-string">'day'</span>] == k,3],
              c=f<span class="org-string">'C{i}'</span>, s=1, alpha=.1, label=k)

ax.set_xlabel(<span class="org-string">'Diffusion map 1'</span>)
ax.set_ylabel(<span class="org-string">'Diffusion map 2'</span>)
ax.set_zlabel(<span class="org-string">'Diffusion map 3'</span>)

<span class="org-variable-name">ax</span> = plt.gcf().add_subplot(122, projection=<span class="org-string">'3d'</span>)
ax.grid(<span class="org-constant">False</span>)
<span class="org-keyword">for</span> i, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(rep2.obs.groupby(<span class="org-string">'day'</span>)):
  ax.scatter(rep2.obsm[<span class="org-string">'X_diffmap'</span>][rep2.obs[<span class="org-string">'day'</span>] == k,1],
              rep2.obsm[<span class="org-string">'X_diffmap'</span>][rep2.obs[<span class="org-string">'day'</span>] == k,2],
              rep2.obsm[<span class="org-string">'X_diffmap'</span>][rep2.obs[<span class="org-string">'day'</span>] == k,3],
              c=f<span class="org-string">'C{i}'</span>, s=1, alpha=.1, label=k)

ax.set_xlabel(<span class="org-string">'Diffusion map 1'</span>)
ax.set_ylabel(<span class="org-string">'Diffusion map 2'</span>)
ax.set_zlabel(<span class="org-string">'Diffusion map 3'</span>)

<span class="org-variable-name">leg</span> = plt.legend(frameon=<span class="org-constant">False</span>, handletextpad=0, markerscale=4, loc=<span class="org-string">'center left'</span>, bbox_to_anchor=(1, .5))
leg.set_title(<span class="org-string">'Day'</span>)
<span class="org-keyword">for</span> h <span class="org-keyword">in</span> leg.legendHandles: 
  h.set_alpha(1)

plt.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/trajectory.org/drop-seq-nmf-diffusion.png" alt="drop-seq-nmf-diffusion.png">
</p>
</div>

<p>
Compute diffusion psuedotime.
</p>

<div class="org-src-container">
<pre class="src src-ipython">np.random.seed(1)
<span class="org-variable-name">dpt</span> = []
<span class="org-keyword">for</span> root <span class="org-keyword">in</span> np.random.choice(rep1.obs[rep1.obs[<span class="org-string">'day'</span>] == 0].index, 200):
  <span class="org-variable-name">rep1.uns</span>[<span class="org-string">'iroot'</span>] = root
  <span class="org-keyword">del</span> rep1.obs[<span class="org-string">'dpt_pseudotime'</span>]
  sc.tl.dpt(rep1, n_dcs=15)
  dpt.append(rep1.obs[<span class="org-string">'dpt_pseudotime'</span>])
<span class="org-variable-name">dpt</span> = np.array(dpt).T
</pre>
</div>

<p>
Look at pseudotime vs. real time.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 8)
<span class="org-variable-name">mean_dpt</span> = dpt.mean(axis=1)
<span class="org-variable-name">order</span> = np.array([i <span class="org-keyword">for</span> k, i <span class="org-keyword">in</span> <span class="org-builtin">sorted</span>([(k, i) <span class="org-keyword">for</span> i, k <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(np.argsort(m))])])
<span class="org-keyword">for</span> i, (k, _) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(rep1.obs.groupby(<span class="org-string">'day'</span>)):
  plt.errorbar(x=mean_dpt[rep1.obs[<span class="org-string">'day'</span>] == k],
           y=order[rep1.obs[<span class="org-string">'day'</span>] == k],
           xerr=dpt[rep1.obs[<span class="org-string">'day'</span>] == k].std(axis=1),
           fmt=<span class="org-string">'o'</span>,
           elinewidth=1,
           markersize=1,
           c=f<span class="org-string">'C{i}'</span>,
           label=k,)
plt.xlabel(<span class="org-string">'Diffusion pseudotime'</span>)
plt.ylabel(<span class="org-string">'Cell'</span>)
<span class="org-variable-name">leg</span> = plt.legend()
leg.set_title(<span class="org-string">'Day'</span>)
</pre>
</div>


<div class="figure">
<p><img src="figure/trajectory.org/drop-seq-nmf-dpt.png" alt="drop-seq-nmf-dpt.png">
</p>
</div>

<p>
Read off gene expression of marker genes (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4546707/">Karakikes et al. 2016</a>) over DPT
trajectories.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">lam</span> = L.dot(nmf.components_)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">markers</span> = [
  <span class="org-comment-delimiter"># </span><span class="org-comment">iPSC</span>
  <span class="org-string">'POU5F1'</span>,  <span class="org-comment-delimiter"># </span><span class="org-comment">Oct-4</span>
  <span class="org-string">'SOX2'</span>,
  <span class="org-string">'NANOG'</span>,
  <span class="org-comment-delimiter"># </span><span class="org-comment">mesoderm formation</span>
  <span class="org-string">'T'</span>,  <span class="org-comment-delimiter"># </span><span class="org-comment">BRY/TBXT</span>
  <span class="org-string">'MIXL1'</span>,
  <span class="org-comment-delimiter"># </span><span class="org-comment">cardiogenic mesoderm</span>
  <span class="org-string">'MESP1'</span>,
  <span class="org-string">'ISL1'</span>,
  <span class="org-string">'KDR'</span>,
  <span class="org-comment-delimiter"># </span><span class="org-comment">cardiac progenitor</span>
  <span class="org-string">'NKX2-5'</span>,
  <span class="org-string">'GATA4'</span>,
  <span class="org-string">'TBX5'</span>,
  <span class="org-string">'MEF2C'</span>,
  <span class="org-string">'HAND1'</span>,
  <span class="org-string">'HAND2'</span>,
  <span class="org-comment-delimiter"># </span><span class="org-comment">cardiomyocyte</span>
  <span class="org-string">'MYL2'</span>,
  <span class="org-string">'MYL7'</span>,
  <span class="org-string">'MYH6'</span>,
  <span class="org-string">'TNNT2'</span>
  ]
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 4)
fig.set_size_inches(11, 3)
<span class="org-keyword">for</span> i, name <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>([<span class="org-string">'POU5F1'</span>, <span class="org-string">'MYL7'</span>, <span class="org-string">'T'</span>, <span class="org-string">'MIXL1'</span>]):
  ax[i].set_yscale(<span class="org-string">'log'</span>)
  <span class="org-keyword">for</span> j, (k, g) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(rep1.obs.groupby(<span class="org-string">'day'</span>)):
    ax[i].scatter(dpt.mean(axis=1)[rep1.obs[<span class="org-string">'day'</span>] == k], lam[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep1'</span>,np.where(rep1.var_names == name)].A.ravel()[rep1.obs[<span class="org-string">'day'</span>] == k], s=1, c=f<span class="org-string">'C{j}'</span>)
  ax[i].set_xlabel(<span class="org-string">'Diffusion pseudotime'</span>)
  ax[i].set_title(name)
ax[0].set_ylabel(<span class="org-string">'Latent expression'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/trajectory.org/drop-seq-nmf-dpt-markers.png" alt="drop-seq-nmf-dpt-markers.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgdef75dc" class="outline-3">
<h3 id="orgdef75dc">NMF pseudotime</h3>
<div class="outline-text-3" id="text-orgdef75dc">
<p>
First, see if we can confidently separate terminal cardiomyocytes.
</p>

<div class="org-src-container">
<pre class="src src-ipython">nmf.components_[:,np.where(rep1.var_names == <span class="org-string">'MYL7'</span>)[0]]
</pre>
</div>

<pre class="example">
array([[3.3862978819e-02],
[3.5804683534e+00],
[0.0000000000e+00],
[1.0358071984e-02],
[0.0000000000e+00],
[1.8786170015e+01],
[0.0000000000e+00],
[0.0000000000e+00],
[6.1099954847e-02],
[0.0000000000e+00]])
</pre>

<p>
Look at linear separability in the largest latent factors.
</p>

<div class="org-src-container">
<pre class="src src-ipython">plt.clf()
plt.gcf().set_size_inches(3, 3)
<span class="org-variable-name">ax</span> = plt.gca()
<span class="org-variable-name">expr</span> = lam[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep1'</span>,np.where(rep1.var_names == <span class="org-string">'MYL7'</span>)[0]].A.ravel()
<span class="org-variable-name">norm</span> = matplotlib.colors.Normalize(vmin=0, vmax=expr.<span class="org-builtin">max</span>())
plt.scatter(1e5 * L[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep1'</span>,1].A, 1e5 * L[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep1'</span>,5].A, s=1, c=colorcet.cm[<span class="org-string">'fire'</span>](norm(expr)))
<span class="org-variable-name">sm</span> = matplotlib.cm.ScalarMappable(norm=norm, cmap=colorcet.cm[<span class="org-string">'fire'</span>])
sm.set_array([])
<span class="org-variable-name">cb</span> = plt.colorbar(sm)
cb.set_label(<span class="org-string">'MYL7'</span>)
plt.xlabel(<span class="org-string">'Factor 1'</span>)
plt.ylabel(<span class="org-string">'Factor 5'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Factor 5')

</pre>

<div class="figure">
<p><img src="figure/trajectory.org/drop-seq-nmf-dpt-terminal.png" alt="drop-seq-nmf-dpt-terminal.png">
</p>
</div>

<p>
Convert to an LDA solution.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">docs</span> = L.A * nmf.components_.<span class="org-builtin">sum</span>(axis=1)
<span class="org-variable-name">topics</span> = nmf.components_ / nmf.components_.<span class="org-builtin">sum</span>(axis=1, keepdims=<span class="org-constant">True</span>)
</pre>
</div>

<p>
Read off the solution. Topic 5 is clearly marker genes for mature
cardiomyocytes.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">idx</span> = np.where(adata.var.index.isin(markers))[0]
plt.clf()
plt.imshow(nmf.components_[:,idx], cmap=colorcet.cm[<span class="org-string">'bmy'</span>])
plt.xticks(np.arange(idx.shape[0]), adata.var.index[idx], rotation=90)
plt.xlabel(<span class="org-string">'Marker gene'</span>)
plt.ylabel(<span class="org-string">'Factor'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Factor')

</pre>

<div class="figure">
<p><img src="figure/trajectory.org/nmf-to-lda-topics-vs-markers.png" alt="nmf-to-lda-topics-vs-markers.png">
</p>
</div>
</div>
</div>

<div id="outline-container-org80fe738" class="outline-3">
<h3 id="org80fe738">Per individual analysis</h3>
<div class="outline-text-3" id="text-org80fe738">
<p>
Try to look at one individual at a time.
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">rep1</span> = adata[adata.obs[<span class="org-string">'ind'</span>] == <span class="org-string">'Rep1'</span>]
<span class="org-variable-name">nmf</span> = skd.NMF(n_components=10, solver=<span class="org-string">'mu'</span>, beta_loss=1)
<span class="org-variable-name">loadings</span> = nmf.fit_transform(rep1.X)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">idx</span> = np.where(rep1.var.index.isin(markers))[0]
plt.clf()
plt.imshow(nmf.components_[:,idx], cmap=colorcet.cm[<span class="org-string">'bmy'</span>])
plt.xticks(np.arange(idx.shape[0]), rep1.var.index[idx], rotation=90)
plt.xlabel(<span class="org-string">'Marker gene'</span>)
plt.ylabel(<span class="org-string">'Factor'</span>)
</pre>
</div>

<pre class="example">
Text(0,0.5,'Factor')

</pre>

<div class="figure">
<p><img src="figure/trajectory.org/rep1-topics-vs-markers.png" alt="rep1-topics-vs-markers.png">
</p>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">docs</span> = loadings * nmf.components_.<span class="org-builtin">sum</span>(axis=1)
<span class="org-variable-name">docs</span> /= docs.<span class="org-builtin">sum</span>(axis=1, keepdims=<span class="org-constant">True</span>)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython">plt.set_cmap(<span class="org-string">'Paired'</span>)
plt.clf()
<span class="org-variable-name">fig</span>, <span class="org-variable-name">ax</span> = plt.subplots(1, 4, sharey=<span class="org-constant">True</span>)
fig.set_size_inches(11, 5)
<span class="org-keyword">for</span> j, (day, a) <span class="org-keyword">in</span> <span class="org-builtin">enumerate</span>(<span class="org-builtin">zip</span>([0, 1, 3, 7], ax)):
  <span class="org-variable-name">prop</span> = np.cumsum(docs[rep1.obs[<span class="org-string">'day'</span>] == day], axis=1)
  <span class="org-variable-name">idx</span> = np.lexsort(docs[rep1.obs[<span class="org-string">'day'</span>] == day].T)
  <span class="org-keyword">for</span> i <span class="org-keyword">in</span> <span class="org-builtin">range</span>(prop.shape[1]):
    <span class="org-keyword">if</span> i &gt; 0:
      <span class="org-variable-name">bot</span> = prop[idx,i - 1]
    <span class="org-keyword">else</span>:
      <span class="org-variable-name">bot</span> = <span class="org-constant">None</span>
    ax[j].bar(np.arange(prop.shape[0]), docs[idx,i], bottom=bot, color=f<span class="org-string">'C{i}'</span>)
fig.tight_layout()
</pre>
</div>


<div class="figure">
<p><img src="figure/trajectory.org/rep1-structure.png" alt="rep1-structure.png">
</p>
</div>
</div>
</div>

<div id="outline-container-orgf6fa745" class="outline-3">
<h3 id="orgf6fa745">Geodesics</h3>
<div class="outline-text-3" id="text-orgf6fa745">
<p>
Estimate the geodesic in this space
(<a href="https://www.springer.com/us/book/9780817634902">https://www.springer.com/us/book/9780817634902</a>, <a href="https://arxiv.org/pdf/1411.7432">https://arxiv.org/pdf/1411.7432</a>
<a href="https://arxiv.org/abs/1710.11379">https://arxiv.org/abs/1710.11379</a>).
</p>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-variable-name">metric</span> = m.components_.dot(m.components_.T)
</pre>
</div>

<div class="org-src-container">
<pre class="src src-ipython"><span class="org-keyword">import</span> scipy.integrate

<span class="org-keyword">def</span> <span class="org-function-name">f</span>(x, y):
  <span class="org-keyword">return</span> np.vstack((y[], -.5 * np.linalg.pinv(metric), np.kron(y[1], y[1])))

<span class="org-keyword">def</span> <span class="org-function-name">bc</span>(ya, yb):
  <span class="org-keyword">return</span> np.array([ya - loadings[start][0], yb - loadings[end][0]])

<span class="org-variable-name">x</span> = np.linspace(0, 1, 5)
<span class="org-variable-name">res</span> = scipy.integrate.solve_bvp(f, bc, x, np.zeros((20, x.size)))
</pre>
</div>

<p>
0 - f554087f-b03a-4574-8613-072c205e5251
</p>
</div>
</div>
</div>
</div>
<div id="postamble" class="status">
<p class="author">Author: Abhishek Sarkar</p>
<p class="date">Created: 2021-08-25 Wed 18:52</p>
<p class="validation"><a href="http://validator.w3.org/check?uri=referer">Validate</a></p>
</div>
</body>
</html>
